release article for nim 1.6.0 (#301)
nim-lang/website: jekyll/_posts/2021-09-25-version-160-released.md

---
title: "Version 1.6.0 released"
author: The Nim Team
---

We are very proud to announce Nim version 1.6, the latest stable release, after a year in the making!
This is by far the biggest release yet, here are some stats:
* 1667 PRs merged (1760 commits)
* 893 issues closed
* 15 new stdlib modules
* new features in >40 stdlib modules, including major improvements to 10 commonly used modules.
* documentation and minor improvements to 170 modules, including 312 new runnable examples
* 280 new nimble packages

> Nim made its first entry in TIOBE index in 2017 at position 129, last year it entered the top-100, and last 2 months it entered the top-50 (https://forum.nim-lang.org/t/8297). We hope this release will reinforce this trend, building on Nim's core strenghs: a practical, compiled systems programming language offering C-like performance and portability, python-like syntax, LISP-like flexibility, strong C, C++, js, python interop and best-in class metaprogramming.

Reviewer: Awesome paragraph, but I feel like that last sentence is far too comma heavy. Any way we can re-word with semi-colons to make it clearer?

> Nim made its first entry in TIOBE index in 2017 at position 129, last year it entered the top-100, and last 2 months it entered the top-50 (https://forum.nim-lang.org/t/8297). We hope this release will reinforce this trend, building on Nim's core strenghs: a practical, compiled systems programming language offering C-like performance and portability, python-like syntax, LISP-like flexibility, strong C, C++, js, python interop and best-in class metaprogramming.

Author: now using semicolons, hopefully easier to parse visually

* made js, VM and nimscript backend more consistent with c backend, allowing more modules to work with those backends, including the imports from `prelude`; the test suite now standardizes on testing stdlib modules on each major backend (c, js, vm).
* support for apple silicon/M1 macs, 32-bit RISC-V, improved support for nodejs backend
* major improvements to the following modules: `system, math, random, json, jsonutils, os, typetraits, wrapnils, lists, hashes` including performance improvements
* deprecated a number of error prone or redundant mis-features

# Installing Nim 1.6
We recommend everyone to upgrade to 1.6:

## New users

Check out if the package manager of your OS already ships version 1.6 or
install it as described [here](https://nim-lang.org/install.html).

## Existing users

If you have installed a previous version of Nim using `choosenim`,
getting Nim 1.6 is as easy as:

```bash
$ choosenim update self
$ choosenim update stable
```

If you don't have `choosenim`, you can follow the same
[install link](https://nim-lang.org/install.html) as above.

## Building from source
```bash
$ git clone https://github.com/nim-lang/Nim
$ cd Nim
$ sh build_all.sh

> ```

Reviewer 2: Why do you have the superuser prompt ($) here and not elsewhere?

Reviewer 3: @quantimnot that's the normal user prompt, superuser prompt is `#`

Author: the other bash snippets indeed don't have `$`, so I've now removed those `$`

# backward compatibility and preview flags

Starting with this release, we've introduced preview flags of the form `-d:nimPreviewX`
(e.g. `-d:nimPreviewFloatRoundtrip`); they allow users to opt-in a new stdlib/compiler behavior
that will likely become default in the next or a future release.
These staging flags are aimed at minimizing backward compatibility issues.
We also introduced opt-out flags of the form `-d:nimLegacyX`,  e.g. `-d:nimLegacyCopyFile`,
for cases where the default was changed to the new behavior,
and an explicit opt-out is needed for some transition period.

# Major new features
With so many new features, pinpointing the most salient ones is a subjective exercise,
but here are a select few:

## strict effects
The effect system was refined and there is a new `.effectsOf` annotation that does
  explicitly what was previously done implicitly. See the manual for details.
  To write code that is portable with older Nim versions, use this idiom:

```nim
when defined(nimHasEffectsOf):
  {.experimental: "strictEffects".}
else:
  {.pragma: effectsOf.}

proc mysort(s: seq; cmp: proc(a, b: T): int) {.effectsOf: cmp.}

> 

Reviewer 2: ```suggestion
```

# instead of:
template fn(a: untyped)
```
This solves a number of long standing issues related to iterators. In particular,
iterable arguments can now be used with MCS, e.g. `iota(3).toSeq` now works.
See PR #17196 for additional details.

## private imports and private field access
- A new import syntax `import foo {.all.}` now allows to import all symbols (public or private)
  from `foo`. This can be useful for testing purposes or for more flexibility in project organization.
- Added a new module `std/importutils`, and an API `privateAccess`, which allows access to private fields
  for an object type in the current scope.

## `nim --eval:cmd`
Added `nim --eval:cmd` to evaluate a command directly:, e.g.: `nim --eval:"echo 1"`.
It defaults to `e` (nimscript) but can also work with other commands, e.g.:
```bash
find . | nim r --eval:'import strutils; for a in stdin.lines: echo a.toUpper'
```
You can now use nim as a calculator, e.g. `nim --eval:'echo 3.1 / (1.2+7)'`.
You can also use it to explore a module's APIs, including private symbols:
```bash
nim --eval:'import os {.all.}; echo weirdTarget'
```
See PR #15687 for more details.

## roundtrip float to string

> `system.addFloat` and `system.$` now can produce string representations of floating point numbers that are minimal in size and that "roundtrip" (via the "Dragonbox" algorithm).

Author: thanks, great suggestion

It is expected that this behavior becomes the new default in upcoming versions, as with other `nimPreviewX` define flags.

## new `std/jsbigints` module
Provides arbitrary precision integers for JavaScript target. See PR #16409.

## new `std/sysrand` module
Cryptographically secure pseudorandom number generator, see PR #16459.

## User defined literals
- Custom numeric literals (e.g. `-128'bignum`) are now supported.
- The unary minus in `-1` is now part of the integer literal, it is now parsed as a single token.
  This implies that edge cases like `-128'i8` finally work correctly.

## new-style concepts
E.g.:
```nim
type
  Comparable = concept # no T, an atom
    proc cmp(a, b: Self): int
```
See PR #15251 for details.

## Dot-like operators
With `-d:nimPreviewDotLikeOps`, dot-like operators (operators starting with `.`, but not with `..`)
now have the same precedence as `.`, so that `a.?b.c` is now parsed as `(a.?b).c` instead of `a.?(b.c)`.
A warning is generated when a dot-like operator is used without `-d:nimPreviewDotLikeOps`.

An important use case is to enable dynamic fields without affecting the builtin `.` operator, e.g. for

> jsffi, json, nimpy.

Author: changed to `std/jsffi`, `std/json`, `pkg/nimpy`

  issues like bug #13063 (which affected error messages)
  for modules importing `std/wrapnils`.
  Added `??.` macro which returns an `Option`.
  `std/wrapnils` can now be used to protect against `FieldDefect` errors in
  case objects, generates optimal code (no overhead compared to manual
  if-else branches), and preserves lvalue semantics which allows modifying
  an expression.

## containers: std/algorithm, std/lists, std/sequtils, std/options, std/packedsets
- Removed the optional `longestMatch` parameter of the `critbits._WithPrefix` iterators (it never worked reliably)
- Added `algorithm.merge`.
- In `lists`: renamed `append` to `add` and retained `append` as an alias;
  added `prepend` and `prependMoved` analogously to `add` and `addMoved`;
  added `remove` for `SinglyLinkedList`s.
- Added new operations for singly- and doubly linked lists: `lists.toSinglyLinkedList`
  and `lists.toDoublyLinkedList` convert from `openArray`s; `lists.copy` implements
  shallow copying; `lists.add` concatenates two lists - an O(1) variation that consumes
  its argument, `addMoved`, is also supplied. 
  See PR #16362, #16536.

- new module: std/packedsets

>   Generalizes std/intsets, see PR #15564.

Author: done

  ```bash
  nim r main # recompiles
  nim r -d:foo main # recompiles
  nim r main # uses cached binary
  nim r main arg1 arg2 # ditto (runtime arguments are irrelevant)
```

- `nim r` now supports cross compilation from unix to windows when specifying `-d:mingw` by using wine,
  e.g.: `nim r --eval:'import os; echo "a" / "b"'` prints `a\b`

- `nim` can compile version 1.4.0 as follows: `nim c --lib:lib --stylecheck:off -d:nimVersion140 compiler/nim`.
  `-d:nimVersion140` is not needed for bootstrapping, only for building 1.4.0 from devel.

- `nim e` now accepts arbitrary file extensions for the nimscript file,
  although `.nims` is still the preferred extension in general.
- The configuration subsystem now allows for `-d:release` and `-d:danger` to work as expected.
  The downside is that these defines now have custom logic that doesn't apply for
  other defines.

## multithreading
- TLS: OSX now uses native TLS (`--tlsEmulation:off`), TLS now works with importcpp non-POD types,
  such types must use `.cppNonPod` and `--tlsEmulation:off`should be used.
- Added `unsafeIsolate` and `extract` to `std/isolation`.
- Added new module: std/tasks

> ## gc

Author: changed to Memory management

- `--gc:arc` now bootstraps (#PR 17342).
- Lots of improvements to gc:arc, gc:orc, refs PR #15697, #16849, #17993.
- `--gc:orc` is now 10% faster than previously for common workloads. If
  you have trouble with its changed behavior, compile with `-d:nimOldOrc`.

> - The `gc:orc` algorithm was refined so that custom container types can participate in the

Reviewer 2: ```suggestion
- The `--gc:orc` algorithm was refined so that custom container types can participate in the
```

---
title: "Version 1.6.0 released"
author: The Nim Team
---

> We are very proud to announce Nim version 1.6, the latest stable release, after a year in the making!
> This is by far the biggest release yet, here are some stats:
> * 1667 PRs merged (1760 commits)
> * 893 issues closed
> * 15 new stdlib modules
> * new features in >40 stdlib modules, including major improvements to 10 commonly used modules.
> * documentation and minor improvements to 170 modules, including 312 new runnable examples
> * 280 new nimble packages

Reviewer 4: ```suggestion
Nim version 1.6 is now officially released!

A year in the making, 1.6 is the latest stable release and by far the largest yet. We're proud of what we, the core team and dedicated volunteers, have accomplished with this milestone. If you're new to Nim, it's a statically typed compiled systems programming language; and version 1.6 makes it all that much more compelling, visit the [Nim homepage](https://nim-lang.org/) to learn more.

Again, none of this would have been possible without the core team, volunteers, and community who's combined efforts resulted in:
* 1667 PRs merged (1760 commits)
* 893 issues closed
* 15 new stdlib modules
* new features in >40 stdlib modules, including major improvements to 10 commonly used modules.
* documentation and minor improvements to 170 modules, including 312 new runnable examples
* 280 new nimble packages
```

If you'd like a bit more flair or "marketing" in the writing.

Author: i took some of your suggestions, but not all:
> If you're new to Nim, it's a statically typed compiled systems programming language; and version 1.6 makes it all that much more compelling, visit the [Nim homepage](https://nim-lang.org/) to learn more.

the problem here is this repeats what's mentioned a few lines below ("building on Nim's core strenghs:...")

> Again, none of this would have been possible without the core team, volunteers, and community who's combined efforts resulted in:

this kind of repeats what's said a few lines earlier ("we, the core team and dedicated volunteers"); see also "Contributors to Nim 1.6" section appearing later

we can iterate ...

a practical, compiled systems programming language; offering C-like performance and portability;
Python-like syntax; LISP-like flexibility; strong C, C++, JS, python interop;
and best-in class metaprogramming.

This release includes improvements in the following areas:
* new language features (user defined literals, private imports, strict effects, `iterable[T]`, new style concepts, dot-like operators, block arguments with optional params)
* new compiler features (`nim --eval:cmd`, custom nimscript extensions, customizable compiler messages)
* major improvements to `--gc:arc`, `--gc:orc`
* correctness and performance of integer and float parsing and rendering in all backends
* significant improvements in error messages, showing useful context
* doc generation logic and documentation, in particular `runnableExamples` now works in more contexts and replaces `code-block`.
* made JS, VM and nimscript backend more consistent with C backend, allowing more modules to work with those backends, including the imports from `std/prelude`; the test suite now standardizes on testing stdlib modules on each major backend (C, JS, VM).

> * support for Apple silicon/M1, 32-bit RISC-V, improved support for NodeJS backend

Author: @Araq should i add CROSSOS to this list (refs https://github.com/nim-lang/Nim/pull/18889)

Author: ended up adding it

- `std/enumutils`
- `std/genasts`
- `std/importutils`
- `std/jsbigints`
- `std/jsfetch`
- `std/jsformdata`
- `std/jsheaders`
- `std/packedsets`
- `std/setutils`
- `std/socketstreams`
- `std/strbasics`
- `std/sysrand`
- `std/tasks`
- `std/tempfiles`
- `std/vmutils`

- Deprecated `std/mersenne`.
- Removed deprecated `std/iup` module from stdlib, it has already moved to
  [nimble](https://github.com/nim-lang/iup).

## New module: `std/setutils`
- Added `setutils.toSet` that can take any iterable and convert it to a built-in `set`,
  if the iterable yields a built-in settable type.
- Added `setutils.fullSet` which returns a full built-in `set` for a valid type.
- Added `setutils.complement` which returns the complement of a built-in `set`.
- Added `setutils.[]=`.

## New module: `std/enumutils`
- Added `genEnumCaseStmt` macro that generates
  case statement to parse string to enum.
- Added `items` for enums with holes.
- Added `symbolName` to return the `enum` symbol name ignoring the human readable name.
- Added `symbolRank` to return the index in which an `enum` member is listed in an enum.

## `system`

> - Added `system.prepareStrMutation` for better support of low

Reviewer 5: Should be `system.prepareMutation`

Author: nice catch, also sent out https://github.com/nim-lang/Nim/pull/18902

```bash
  nim r main # recompiles
  nim r -d:foo main # recompiles
  nim r main # uses cached binary
  nim r main arg1 arg2 # ditto (runtime arguments are irrelevant)
```

- `nim r` now supports cross compilation from unix to windows when specifying `-d:mingw` by using Wine,
  e.g.: `nim r --eval:'import os; echo "a" / "b"'` prints `a\b`

- `nim` can compile version 1.4.0 as follows: `nim c --lib:lib --stylecheck:off -d:nimVersion140 compiler/nim`.
  `-d:nimVersion140` is not needed for bootstrapping, only for building 1.4.0 from devel.

- `nim e` now accepts arbitrary file extensions for the nimscript file,
  although `.nims` is still the preferred extension in general.
- The configuration subsystem now allows for `-d:release` and `-d:danger` to work as expected.
  The downside is that these defines now have custom logic that doesn't apply for
  other defines.

## Multithreading
- TLS: OSX now uses native TLS (`--tlsEmulation:off`), TLS now works with importcpp non-POD types,
  such types must use `.cppNonPod` and `--tlsEmulation:off`should be used.
- Added `unsafeIsolate` and `extract` to `std/isolation`.

> - Added new module: `std/tasks`

Reviewer 4: ```suggestion
- Added `std/tasks`, a new module containing primitives for creating parallel programs.
```

## Memory management
- `--gc:arc` now bootstraps (PR [#17342](https://github.com/nim-lang/Nim/pull/17342)).
- Lots of improvements to `gc:arc`, `gc:orc`, refs PR [#15697](https://github.com/nim-lang/Nim/pull/15697), [#16849](https://github.com/nim-lang/Nim/pull/16849), [#17993](https://github.com/nim-lang/Nim/pull/17993).
- `--gc:orc` is now 10% faster than previously for common workloads. If
  you have trouble with its changed behavior, compile with `-d:nimOldOrc`.
- The `--gc:orc` algorithm was refined so that custom container types can participate in the
  cycle collection process. See the documentation of `=trace` for more details.
- On embedded devices `malloc` can now be used instead of `mmap` via `-d:nimAllocPagesViaMalloc`.
  This is only supported for `--gc:orc` or `--gc:arc`.

Compatibility notes:

> - `--newruntime` and `--refchecks` are deprecated.

Reviewer 4: ```suggestion
- `--newruntime` and `--refchecks` are deprecated,
  use `--gc:arc`, --gc:orc`, or `--gc:none` as appropriate instead.
```

```
- docgen now renders deprecated and other pragmas.

- `runnableExamples` now works with templates and nested templates.
- `runnableExamples: "-r:off"` now works for examples that should compile but not run. 
- `runnableExamples` now renders code verbatim, and produces correct code in all cases.
- docgen now shows correct, canonical import paths in docs.
- docgen now shows all routines in sidebar, and the proc signature is now shown in sidebar.

## Effects and checks
- Significant improvement to error messages involving effect mismatches

- There is a new `cast` section `{.cast(uncheckedAssign).}: body` that disables some
  compiler checks regarding `case objects`. This allows serialization libraries
  to avoid ugly, non-portable solutions. See RFC [#407](https://github.com/nim-lang/RFCs/issues/407).
  for more details.

## Tools
- major improvements to `nimgrep`, see PR [#15612
](https://github.com/nim-lang/Nim/pull/15612).
- `fusion` is now un-bundled from Nim, `./koch fusion` will
  install it via Nimble at a fixed hash.

- `testament`: added `nimoutFull: bool` spec to compare full output of compiler
  instead of a subset; many bugfixes to testament.

> - Added `atlas` helper tool.

Reviewer 4: ```suggestion
- Added `atlas` a tool for automating workflows involing "cloning" a nimble package and dependencies recursively
```

* new compiler features (`nim --eval:cmd`, custom nimscript extensions, customizable compiler messages)
* major improvements to `--gc:arc`, `--gc:orc`
* correctness and performance of integer and float parsing and rendering in all backends
* significant improvements in error messages, showing useful context
* doc generation logic and documentation, in particular `runnableExamples` now works in more contexts and replaces `code-block`.
* made JS, VM and nimscript backend more consistent with C backend, allowing more modules to work with those backends, including the imports from `std/prelude`; the test suite now standardizes on testing stdlib modules on each major backend (C, JS, VM).
* support for Apple silicon/M1, 32-bit RISC-V, CROSSOS, improved support for NodeJS backend
* major improvements to the following modules: `system, math, random, json, jsonutils, os, typetraits, wrapnils, lists, hashes` including performance improvements
* deprecated a number of error prone or redundant mis-features

# Installing Nim 1.6

> We recommend everyone to upgrade to 1.6:

Reviewer 4: ```suggestion
We recommend everyone upgrade to 1.6:
```

Author: "we recommend everyone to upgrade" gives way more google hits :)

Example:
```nim
import std/sysrand
assert urandom(1234) != urandom(1234) # unlikely to fail in practice
```

## New module: `std/tempfiles`
Allows creating temporary files and directories, see PR [#17361](https://github.com/nim-lang/Nim/pull/17361) and followups.
```nim
import std/tempfiles
let tmpPath = genTempPath("prefix", "suffix.log", "/tmp/")
# tmpPath looks like: /tmp/prefixpmW1P2KLsuffix.log

let dir = createTempDir("tmpprefix_", "_end")
# created dir looks like: getTempDir() / "tmpprefix_YEl9VuVj_end"

let (cfile, path) = createTempFile("tmpprefix_", "_end.tmp")
# path looks like: getTempDir() / "tmpprefix_FDCIRZA0_end.tmp"
cfile.write "foo"
cfile.setFilePos 0
assert readAll(cfile) == "foo"
close cfile
assert readFile(path) == "foo"
```

## User defined literals
- Custom numeric literals (e.g. `-128'bignum`) are now supported.
- The unary minus in `-1` is now part of the integer literal, it is now parsed as a single token.
  This implies that edge cases like `-128'i8` finally work correctly.
Example:
```nim
func `'big`*(num: cstring): JsBigInt {.importjs: "BigInt(#)".}
assert 0xffffffffffffffff'big == (1'big shl 64'big) - 1'big
```

## New-style concepts

> E.g.:

Reviewer 4: ```suggestion
Example:
```

Constistent with other text.

```nim
type
  Comparable = concept # no T, an atom
    proc cmp(a, b: Self): int
```

> See PR [#15251](https://github.com/nim-lang/Nim/pull/15251) for details.

Reviewer 5: Are the new concepts worth mentioning? The linked PR says it was merged to prevent bitrotting not because it was complete and there doesn't seem to be much testing around it.

Author: I'm on the fence; how about I move it to the section "# Other new features" instead of "# Major new features" ?

Author: done

---
title: "Version 1.6.0 released"
author: The Nim Team
---

Nim version 1.6 is now officially released!

A year in the making, 1.6 is the latest stable release and by far the largest yet.
We're proud of what we, the core team and dedicated volunteers, have accomplished with this milestone.
Here are some stats:
* 1667 PRs merged (1760 commits)
* 893 issues closed
* 15 new stdlib modules
* new features in >40 stdlib modules, including major improvements to 10 commonly used modules.
* documentation and minor improvements to 170 modules, including 312 new runnable examples
* 280 new nimble packages

Nim made its first entry in TIOBE index in 2017 at position 129,
last year it entered the top-100, and last 2 months it entered the top-50 (https://forum.nim-lang.org/t/8297).

> We hope this release will reinforce this trend, building on Nim's core strenghs:

Reviewer: ```suggestion
We hope this release will reinforce this trend, building on Nim's core strengths:
```

a practical, compiled systems programming language; offering C-like performance and portability;

> Python-like syntax; LISP-like flexibility; strong C, C++, JS, python interop;

Reviewer: ```suggestion
Python-like syntax; LISP-like flexibility; strong C, C++, JS, Python interop;
```

and best-in class metaprogramming.

This release includes improvements in the following areas:
* new language features (user defined literals, private imports, strict effects, `iterable[T]`, new style concepts, dot-like operators, block arguments with optional params)
* new compiler features (`nim --eval:cmd`, custom nimscript extensions, customizable compiler messages)
* major improvements to `--gc:arc`, `--gc:orc`
* correctness and performance of integer and float parsing and rendering in all backends
* significant improvements in error messages, showing useful context

> * doc generation logic and documentation, in particular `runnableExamples` now works in more contexts and replaces `code-block`.

Reviewer: This implies `code-block` is deprated, is that the case?

Author: if we can merge this PR https://github.com/nim-lang/Nim/pull/18604, then I could change the statement to "now works in every context" instead of "more", and then code-block for nim code can indeed be considered deprecated within nim files (but could have occasional use for non-nim code within nim files, as well as its non-deprecated use in rst files)

Reviewer 6: > This implies code-block is deprated, is that the case?

It is not the case.

Reviewer: we might want to reword this then: just remove "and replaces `code-block`."

* deprecated a number of error prone or redundant mis-features

# Installing Nim 1.6
We recommend everyone to upgrade to 1.6:

## New users

Check out if the package manager of your OS already ships version 1.6 or
install it as described [here](https://nim-lang.org/install.html).

## Existing users

If you have installed a previous version of Nim using `choosenim`,
getting Nim 1.6 is as easy as:

```bash
choosenim update self
choosenim update stable
```

If you don't have `choosenim`, you can follow the same
[install link](https://nim-lang.org/install.html) as above.

## Building from source
```bash
git clone https://github.com/nim-lang/Nim
cd Nim
sh build_all.sh
```
The last command can be re-run after pulling new commits.
Note that the `csources` repo used was changed to `csources_v1`, the new setup is designed to be forward and backward compatible.

## Building from a CI setup
We now have an API to be used in CI which abstracts the implementation details: `. ci/funs.sh && nimBuildCsourcesIfNeeded`; in fact all the existing CI pipelines have been refactored to use this, see

> [#17815](https://github.com/nim-lang/Nim/pull/17815).

Reviewer: Is this really relevant to Nim users? Nim users shouldn't be building Nim in their CI and instead be grabbing a nightly or a Docker image with a specific Nim version.

I would recommend removing it.

Author: I've now re-phrased this section and mentioned nightlies and docker as alternatives for CI, but this is IMO still relevant; in fact https://github.com/search?p=1&q=nimBuildCsourcesIfNeeded&type=Code shows several use cases of `nimBuildCsourcesIfNeeded`; there are always cases where building from source is needed even in CI.

this is better than what was previously done (hardcoding instructions to a fixed hash which had all the problems discussed in https://github.com/nim-lang/Nim/pull/17815)

# Contributors to Nim 1.6

Many thanks to our recurring and new [contributors](https://github.com/nim-lang/Nim/graphs/contributors?from=2020-10-16&to=2021-09-25&type=c),
Nim is a community driven collaborative effort that welcomes all contributions, big or small.

# Backward compatibility and preview flags

> Starting with this release, we've introduced preview flags of the form `-d:nimPreviewX`
> (e.g. `-d:nimPreviewFloatRoundtrip`); they allow users to opt-in a new stdlib/compiler behavior
> that will likely become default in the next or a future release.
> These staging flags are aimed at minimizing backward compatibility issues.
> We also introduced opt-out flags of the form `-d:nimLegacyX`,  e.g. `-d:nimLegacyCopyFile`,
> for cases where the default was changed to the new behavior,
> and an explicit opt-out is needed for some transition period.

Reviewer: This is a hefty paragraph, consider splitting up:
```suggestion
Starting with this release, we've introduced preview flags of the form `-d:nimPreviewX`
(e.g. `-d:nimPreviewFloatRoundtrip`); they allow users to opt-in a new stdlib/compiler behavior
that will likely become default in the next or a future release.
These staging flags are aimed at minimizing backward compatibility issues.

We also introduced opt-out flags of the form `-d:nimLegacyX`,  e.g. `-d:nimLegacyCopyFile`,
for cases where the default was changed to the new behavior,
and an explicit opt-out is needed for some transition period.
```

# Major new features
With so many new features, pinpointing the most salient ones is a subjective exercise,
but here are a select few:

## Strict effects
The effect system was refined and there is a new `.effectsOf` annotation that does

>   explicitly what was previously done implicitly. See the manual for details.

Reviewer: Link here for the manual would be awesome :)

Author: done

  {.experimental: "strictEffects".}
else:
  {.pragma: effectsOf.}

proc mysort(s: seq; cmp: proc(a, b: T): int) {.effectsOf: cmp.}
```

To enable the new effect system, use --experimental:strictEffects. See [#18777](https://github.com/nim-lang/Nim/pull/18777) and RFC [#408](https://github.com/nim-lang/RFCs/issues/408).

## `iterable[T]`
Added `iterable[T]` type class to match called iterators, which enables writing:
```nim
template fn(a: iterable) # or template fn[T](a: iterable[T])
# instead of:
template fn(a: untyped)
```
This solves a number of long standing issues related to iterators. In particular,
iterable arguments can now be used with MCS, e.g. `iota(3).toSeq` now works.
See PR [#17196](https://github.com/nim-lang/Nim/pull/17196) for additional details.

## Private imports and private field access
- A new import syntax `import foo {.all.}` now allows to import all symbols (public or private)
  from `foo`. This can be useful for testing purposes or for more flexibility in project organization.
- Added a new module `std/importutils`, and an API `privateAccess`, which allows access to private fields
  for an object type in the current scope.
Example:
```nim

> from system as system2 {.all.} import ThisIsSystem

Reviewer: Is this valid syntax? the `import ThisIsSystem` looks wrong

Author: actually there was a typo in what i wrote, I've now corrected it and also added an example for `privateImport` (note that the problem wasn't `import ThisIsSystem` but the placement of `{.all.}`; `from system {.all.} as system2 import nil` is valid and so is `from system {.all.} as system2 import ThisIsSystem `

- Added `hasClosure`.
- Added `pointerBase` to return `T` for `ref T | ptr T`.
- Added `compilesettings.SingleValueSetting.libPath`.

## networking: `std/net`, `std/asyncnet`, `std/htmlgen`, `std/httpclient`, `std/asyncdispatch`, `std/asynchttpserver`, `std/httpcore`
- Fixed buffer overflow bugs in `std/net`.
- Exported `sslHandle` from `std/net` and `std/asyncnet`.
- Added `hasDataBuffered` to `std/asyncnet`.
- various functions in `std/httpclient` now accept `url` of type `Uri`. Moreover `request` function's
  `httpMethod` argument of type `string` was deprecated in favor of `HttpMethod` `enum` type; see [#15919](https://github.com/nim-lang/Nim/pull/15919).
- Added `asyncdispatch.activeDescriptors` that returns the number of currently
  active async event handles/file descriptors.
- Added `getPort` to `std/asynchttpserver` to resolve OS-assigned `Port(0)`;
  this is usually recommended instead of hardcoding ports which can lead to "Address already in use" errors.
- Fixed premature garbage collection in `std/asyncdispatch`, when a stacktrace override is in place.
- Added `httpcore.is1xx` and missing HTTP codes.

> - Added `htmlgen.portal` for [making "SPA style" pages using HTML only](https://web.dev/hands-on-portals).

Reviewer: nit: this isn't related to networking IMO

Author: yes but where else do you suggest so it doesn't in a section of its own?

Reviewer: Meh, I guess "networking" is fine.

- TLS: OSX now uses native TLS (`--tlsEmulation:off`), TLS now works with importcpp non-POD types,
  such types must use `.cppNonPod` and `--tlsEmulation:off`should be used.
- Added `unsafeIsolate` and `extract` to `std/isolation`.
- Added `std/tasks`, a new module containing primitives for creating parallel programs.

## Memory management
- `--gc:arc` now bootstraps (PR [#17342](https://github.com/nim-lang/Nim/pull/17342)).
- Lots of improvements to `gc:arc`, `gc:orc`, refs PR [#15697](https://github.com/nim-lang/Nim/pull/15697), [#16849](https://github.com/nim-lang/Nim/pull/16849), [#17993](https://github.com/nim-lang/Nim/pull/17993).
- `--gc:orc` is now 10% faster than previously for common workloads. If
  you have trouble with its changed behavior, compile with `-d:nimOldOrc`.
- The `--gc:orc` algorithm was refined so that custom container types can participate in the
  cycle collection process. See the documentation of `=trace` for more details.
- On embedded devices `malloc` can now be used instead of `mmap` via `-d:nimAllocPagesViaMalloc`.
  This is only supported for `--gc:orc` or `--gc:arc`.

Compatibility notes:
- `--newruntime` and `--refchecks` are deprecated,

>   use `--gc:arc`, --gc:orc`, or `--gc:none` as appropriate instead.

Reviewer: Missing a \` here
```suggestion
  use `--gc:arc`, `--gc:orc`, or `--gc:none` as appropriate instead.
```

```
- docgen now renders deprecated and other pragmas.

- `runnableExamples` now works with templates and nested templates.
- `runnableExamples: "-r:off"` now works for examples that should compile but not run. 
- `runnableExamples` now renders code verbatim, and produces correct code in all cases.
- docgen now shows correct, canonical import paths in docs.
- docgen now shows all routines in sidebar, and the proc signature is now shown in sidebar.

## Effects and checks
- Significant improvement to error messages involving effect mismatches

- There is a new `cast` section `{.cast(uncheckedAssign).}: body` that disables some
  compiler checks regarding `case objects`. This allows serialization libraries
  to avoid ugly, non-portable solutions. See RFC [#407](https://github.com/nim-lang/RFCs/issues/407).
  for more details.

## Tools
- major improvements to `nimgrep`, see PR [#15612
](https://github.com/nim-lang/Nim/pull/15612).
- `fusion` is now un-bundled from Nim, `./koch fusion` will
  install it via Nimble at a fixed hash.

- `testament`: added `nimoutFull: bool` spec to compare full output of compiler
  instead of a subset; many bugfixes to testament.

> - Added `atlas` a tool for automating workflows involing "cloning" a nimble package and dependencies recursively

Reviewer: ```suggestion
- Added `atlas` a tool for automating workflows involving "cloning" a nimble package and dependencies recursively
```

Reviewer: If atlas is for internal use then I would just avoid mentioning it

Author: based on the discussions from this PR https://github.com/nim-lang/Nim/pull/18756#issuecomment-908088929, I have to agree with you; I've now removed the mention

---
title: "Version 1.6.0 released"
author: The Nim Team
---

Nim version 1.6 is now officially released!

A year in the making, 1.6 is the latest stable release and by far the largest yet.
We're proud of what we, the core team and dedicated volunteers, have accomplished with this milestone.
Here are some stats:
* 1667 PRs merged (1760 commits)
* 893 issues closed
* 15 new stdlib modules
* new features in >40 stdlib modules, including major improvements to 10 commonly used modules.
* documentation and minor improvements to 170 modules, including 312 new runnable examples
* 280 new nimble packages

Nim made its first entry in TIOBE index in 2017 at position 129,
last year it entered the top-100, and last 2 months it entered the top-50 (https://forum.nim-lang.org/t/8297).
We hope this release will reinforce this trend, building on Nim's core strengths:
a practical, compiled systems programming language; offering C-like performance and portability;
Python-like syntax; LISP-like flexibility; strong C, C++, JS, Python interop;

> and best-in class metaprogramming.

Author: thinking of something like this, improvements/suggestions welcome:

Did we mention fast compile times (8s to recompile nim vs XXX for rust), compiles and runs wherever you have a C compiler [gameboy kickstarter](https://forum.nim-lang.org/t/8375), and yes, even write [malware](https://thehackernews.com/2021/03/researchers-spotted-malware-written-in.html).

Reviewer 5: I think the Go program has fewer system dependencies so may not be a fair comparison.

Reviewer 3: IMO it's better to avoid mentioning malware, but that's just my view

Reviewer 7: 2MB is pretty fair given that's caused by Go's language design. Go's reflection mechanism means you cannot optimize away unused routines.

Reviewer 3: @deech even if you link nim hello world binary statically, it'll surely be under 1MB, probably under 200KB

Reviewer: I agree, let's not mention malware, especially since many people are getting false positives alerts about Nim being malware...

About the rest, it sounds weirdly worded. You should instead just have a section where exciting projects are mentioned, in particular the kickstarter.

Author: > I agree, let's not mention malware

yes, this was a long shot; I'm not mentioning it.

I've added a section "Why use Nim?", please provide feedback

getting Nim 1.6 is as easy as:

```bash
choosenim update self
choosenim update stable
```

If you don't have `choosenim`, you can follow the same
[install link](https://nim-lang.org/install.html) as above.

## Building from source
```bash
git clone https://github.com/nim-lang/Nim
cd Nim
sh build_all.sh
```
The last command can be re-run after pulling new commits.
Note that the `csources` repo used was changed to `csources_v1`, the new setup is designed to be forward and backward compatible.

## Building from a CI setup
We now have bash APIs to (re-)build Nim from source which hide implementation details, for example:
`. ci/funs.sh && nimBuildCsourcesIfNeeded`.
This can be useful for CI when alternatives (using nightly builds or a Docker image) are not suitable;
in fact all the existing CI pipelines have been refactored to use this, see
[#17815](https://github.com/nim-lang/Nim/pull/17815).

# Contributors to Nim 1.6

Many thanks to our recurring and new [contributors](https://github.com/nim-lang/Nim/graphs/contributors?from=2020-10-16&to=2021-09-25&type=c),
Nim is a community driven collaborative effort that welcomes all contributions, big or small.

> # Backward compatibility and preview flags

Reviewer 8: Is there a list with all such preview/legacy flags? If so, it should be linked here, otherwise, we should probably make one.

Author: I've finally added this, see text below "Here's the list of these flags"

> Starting with this release, we've introduced preview flags of the form `-d:nimPreviewX`
> (e.g. `-d:nimPreviewFloatRoundtrip`); they allow users to opt-in a new stdlib/compiler behavior
> that will likely become default in the next or a future release.
> These staging flags are aimed at minimizing backward compatibility issues.

Reviewer 8: ```suggestion
Starting with this release, we've introduced preview flags of the form `-d:nimPreviewX`
(e.g. `-d:nimPreviewFloatRoundtrip`), which allow users to opt-in to new stdlib/compiler behavior
that will likely become the default in the next or a future release.
These staging flags are aimed at minimizing backward compatibility issues.
```

> We also introduced opt-out flags of the form `-d:nimLegacyX`,  e.g. `-d:nimLegacyCopyFile`,
> for cases where the default was changed to the new behavior,
> and an explicit opt-out is needed for some transition period.

Reviewer 8: ```suggestion
We also introduced opt-out flags of the form `-d:nimLegacyX`,  e.g. `-d:nimLegacyCopyFile`,
for cases where the default was changed to the new behavior.
For a transition period, these flags can be used to get the old behaviour.
```
This makes it clearer what is opt-out and what the transition period is for imo.

# Major new features
With so many new features, pinpointing the most salient ones is a subjective exercise,
but here are a select few:

## Strict effects
The effect system was refined and there is a new `.effectsOf` annotation that does

> explicitly what was previously done implicitly. See [manual](https://nim-lang.github.io/Nim/manual.html#effect-system-effectsof-annotation) for details.

Reviewer 8: ```suggestion
explicitly what was previously done implicitly. See the [manual](https://nim-lang.github.io/Nim/manual.html#effect-system-effectsof-annotation) for more details.
```

To write code that is portable with older Nim versions, use this idiom:

```nim
when defined(nimHasEffectsOf):
  {.experimental: "strictEffects".}
else:
  {.pragma: effectsOf.}

proc mysort(s: seq; cmp: proc(a, b: T): int) {.effectsOf: cmp.}
```

> To enable the new effect system, use --experimental:strictEffects. See [#18777](https://github.com/nim-lang/Nim/pull/18777) and RFC [#408](https://github.com/nim-lang/RFCs/issues/408).

Reviewer 8: ```suggestion
To enable the new effect system, compile with `--experimental:strictEffects`. See also [#18777](https://github.com/nim-lang/Nim/pull/18777) and RFC [#408](https://github.com/nim-lang/RFCs/issues/408).
```

## `iterable[T]`

> Added `iterable[T]` type class to match called iterators, which enables writing:

Reviewer 8: ```suggestion
The `iterable[T]` type class was added to match called iterators, which enables writing:
```

```bash
find . | nim r --eval:'import strutils; for a in stdin.lines: echo a.toUpper'
```
You can now use nim as a calculator, e.g. `nim --eval:'echo 3.1 / (1.2+7)'`.
You can also use it to explore a module's APIs, including private symbols:
```bash
nim --eval:'import os {.all.}; echo weirdTarget'
```
See PR [#15687](https://github.com/nim-lang/Nim/pull/15687) for more details.

## Round-trip float to string
`system.addFloat` and `system.$` now can produce string representations of floating point numbers that are minimal in size and possess round-trip and correct rounding guarantees (via the [Dragonbox](https://raw.githubusercontent.com/jk-jeon/dragonbox/master/other_files/Dragonbox.pdf) algorithm).
This currently has to be enabled via `-d:nimPreviewFloatRoundtrip`.
It is expected that this behavior becomes the new default in upcoming versions, as with other `nimPreviewX` define flags.

Example:
```nim
from math import round
let a = round(9.779999999999999, 2)
assert a == 9.78
echo a # with `-d:nimPreviewFloatRoundtrip`: 9.78, like in python3 (instead of  9.779999999999999)
```

## New `std/jsbigints` module

> Provides arbitrary precision integers for JS target. See PR [#16409](https://github.com/nim-lang/Nim/pull/16409).

Reviewer 8: ```suggestion
Provides arbitrary precision integers for the JS target. See PR [#16409](https://github.com/nim-lang/Nim/pull/16409).
```

Example:
```nim
import std/jsbigints

> assert big"2" ** big"64" == big"18446744073709551616"

Reviewer 8: `2'big` doesn't work yet, does it? That would be a way to also show the new user-defined number literals.

Author: it does actually, i added support for it in https://github.com/nim-lang/Nim/pull/17707; but good point, I've now added examples demonstrating it

```

## New `std/sysrand` module
Cryptographically secure pseudorandom number generator, see PR [#16459](https://github.com/nim-lang/Nim/pull/16459).
Example:
```nim
import std/sysrand
assert urandom(1234) != urandom(1234) # unlikely to fail in practice
```

## New module: `std/tempfiles`
Allows creating temporary files and directories, see PR [#17361](https://github.com/nim-lang/Nim/pull/17361) and followups.
```nim
import std/tempfiles
let tmpPath = genTempPath("prefix", "suffix.log", "/tmp/")
# tmpPath looks like: /tmp/prefixpmW1P2KLsuffix.log

let dir = createTempDir("tmpprefix_", "_end")
# created dir looks like: getTempDir() / "tmpprefix_YEl9VuVj_end"

let (cfile, path) = createTempFile("tmpprefix_", "_end.tmp")
# path looks like: getTempDir() / "tmpprefix_FDCIRZA0_end.tmp"
cfile.write "foo"
cfile.setFilePos 0
assert readAll(cfile) == "foo"
close cfile
assert readFile(path) == "foo"
```

## User defined literals

> - Custom numeric literals (e.g. `-128'bignum`) are now supported.
> - The unary minus in `-1` is now part of the integer literal, it is now parsed as a single token.
>   This implies that edge cases like `-128'i8` finally work correctly.

Reviewer 8: ```suggestion
Custom numeric literals (e.g. `-128'bignum`) are now supported.
Additionally, the unary minus in `-1` is now part of the integer literal, i.e. it is now parsed as a single token.
This implies that edge cases like `-128'i8` finally work correctly.
```

Example:
```nim
func `'big`*(num: cstring): JsBigInt {.importjs: "BigInt(#)".}
assert 0xffffffffffffffff'big == (1'big shl 64'big) - 1'big
```

## Dot-like operators
With `-d:nimPreviewDotLikeOps`, dot-like operators (operators starting with `.`, but not with `..`)
now have the same precedence as `.`, so that `a.?b.c` is now parsed as `(a.?b).c` instead of `a.?(b.c)`.
A warning is generated when a dot-like operator is used without `-d:nimPreviewDotLikeOps`.

An important use case is to enable dynamic fields without affecting the builtin `.` operator, e.g. for
`std/jsffi`, `std/json`, `pkg/nimpy`. Example:
```nim
import std/json
template `.?`(a: JsonNode, b: untyped{ident}): JsonNode =
  a[astToStr(b)]
let j = %*{"a1": {"a2": 10}}
assert j.?a1.?a2.getInt == 10
```

> ## Block arguments now support optional params
> This solves a major pain point for routines accepting block params, see PR #18631 for details:

Reviewer 8: ```suggestion
## Block arguments now support optional parameters
This solves a major pain point for routines accepting block parameters, see PR #18631 for details:
```

```nim
template fn(a = 1, b = 2, body) = discard
fn(1, 2): # already works
  bar
fn(a = 1): # now works
  bar
```

> ditto with multiple block args via do:

Reviewer 8: ```suggestion
ditto with multiple block arguments via `do`:
```

  if the iterable yields a built-in settable type.
- Added `setutils.fullSet` which returns a full built-in `set` for a valid type.
- Added `setutils.complement` which returns the complement of a built-in `set`.
- Added `setutils.[]=`.

## New module: `std/enumutils`
- Added `genEnumCaseStmt` macro that generates
  case statement to parse string to enum.
- Added `items` for enums with holes.
- Added `symbolName` to return the `enum` symbol name ignoring the human readable name.
- Added `symbolRank` to return the index in which an `enum` member is listed in an enum.

## `system`
- Added `system.prepareMutation` for better support of low
  level `moveMem`, `copyMem` operations for `gc:orc`'s copy-on-write string
  implementation.
- `system.addEscapedChar` now renders `\r` as `\r` instead of `\c`, to be compatible
  with most other languages.
- Added `cmpMem` to `system`.
- `doAssertRaises` now correctly handles foreign exceptions.
- `addInt` now supports unsigned integers

Compatibility notes:
- `system.delete` had a most surprising behavior when the index passed to it was out of
  bounds (it would delete the last entry then). Compile with `-d:nimStrictDelete` so

>   that an index error is produced instead. But be aware that your code might depend on

Reviewer 8: ```suggestion
  that an index error is produced instead. Be aware however that your code might depend on
```

  this quirky behavior so a review process is required on your part before you can
  use `-d:nimStrictDelete`. To make this review easier, use the `-d:nimAuditDelete`
  switch, it pretends that `system.delete` is deprecated so that it is easier to see
  where it was used in your code.

  `-d:nimStrictDelete` will become the default in upcoming versions.
- `cuchar` is now deprecated as it aliased `char` where arguably it should have aliased `uint8`.
  Please use `char` or `uint8` instead.

> - `repr` now doesn't insert trailing newline; previous behavior was very inconsistent,
>   see [#16034](https://github.com/nim-lang/Nim/pull/16034). Use `-d:nimLegacyReprWithNewline` for previous behavior. `repr` now also

Reviewer 8: ```suggestion
- `repr` now doesn't insert trailing newlines; the previous behavior was very inconsistent,
  see [#16034](https://github.com/nim-lang/Nim/pull/16034). Use `-d:nimLegacyReprWithNewline` for the previous behavior. `repr` now also
```

  renders ASTs correctly for user defined literals, setters, `do`, etc.
- Deprecated `any`. See RFC [#281](https://github.com/nim-lang/RFCs/issues/281).
- The unary slice `..b` was deprecated, use `0..b` instead.

## `std/math`
- Added `almostEqual` for comparing two float values using a machine epsilon.
- Added `clamp` which allows using a `Slice` to clamp to a value.
- Added `ceilDiv` for round up integer division.
- Added `isNaN`.
- Added `copySign`.
- Added `euclDiv` and `euclMod`.
- Added `signbit`.
- Added `frexp` overload procs. Deprecated `c_frexp`, use `frexp` instead.

Compatibility notes:

> - `math.round` now is rounded "away from zero" in JS backend which is consistent
>   with other backends. See [#9125](https://github.com/nim-lang/Nim/pull/9125). Use `-d:nimLegacyJsRound` for previous behavior.

Reviewer 8: ```suggestion
- `math.round` now rounds "away from zero" in the JS backend, which is consistent
  with other backends. See [#9125](https://github.com/nim-lang/Nim/pull/9125). Use `-d:nimLegacyJsRound` for the previous behavior.
```

## Random number generators: `std/random`, `std/sysrand`, `std/oids`
- Added `randState` template that exposes the default random number generator.
  Useful for library authors.
- Added `initRand()` overload with no argument which uses the current time as a seed.
- `initRand(seed)` now allows `seed == 0`.

> - Added `std/sysrand` module to get random numbers from a secure source

Reviewer 8: ```suggestion
- Added `std/sysrand` module to get random numbers from a secure source.
```

  instead of raising (as in 1.4) or giving wrong results (as in 1.2).
- `std/jsonutils` now handles `cstring` (including as Table key), and `set`.
- added `jsonutils.jsonTo` overload with `opt = Joptions()` param.
- `jsonutils.toJson` now supports customization via `ToJsonOptions`.
- `std/json`, `std/jsonutils` now support round-trip serialization when `-d:nimPreviewFloatRoundtrip` is used.

## `std/typetraits`, `std/compilesettings`
- `distinctBase` now is identity instead of error for non distinct types.
- `distinctBase` now allows controlling whether to be recursive or not.
- Added `enumLen` to return the number of elements in an enum.
- Added `HoleyEnum` for enums with holes, `OrdinalEnum` for enums without holes.
- Added `hasClosure`.
- Added `pointerBase` to return `T` for `ref T | ptr T`.
- Added `compilesettings.SingleValueSetting.libPath`.

## networking: `std/net`, `std/asyncnet`, `std/htmlgen`, `std/httpclient`, `std/asyncdispatch`, `std/asynchttpserver`, `std/httpcore`
- Fixed buffer overflow bugs in `std/net`.
- Exported `sslHandle` from `std/net` and `std/asyncnet`.
- Added `hasDataBuffered` to `std/asyncnet`.

> - various functions in `std/httpclient` now accept `url` of type `Uri`. Moreover `request` function's

Reviewer 8: ```suggestion
- Various functions in `std/httpclient` now accept `url` of type `Uri`. Moreover `request` function's
```

  `httpMethod` argument of type `string` was deprecated in favor of `HttpMethod` `enum` type; see [#15919](https://github.com/nim-lang/Nim/pull/15919).
- Added `asyncdispatch.activeDescriptors` that returns the number of currently
  active async event handles/file descriptors.
- Added `getPort` to `std/asynchttpserver` to resolve OS-assigned `Port(0)`;
  this is usually recommended instead of hardcoding ports which can lead to "Address already in use" errors.
- Fixed premature garbage collection in `std/asyncdispatch`, when a stacktrace override is in place.
- Added `httpcore.is1xx` and missing HTTP codes.
- Added `htmlgen.portal` for [making "SPA style" pages using HTML only](https://web.dev/hands-on-portals).

Compatibility notes:

> - On Windows the SSL library now checks for valid certificates.

Reviewer 8: ```suggestion
- On Windows, the SSL library now checks for valid certificates.
```

  It uses the `cacert.pem` file for this purpose which was extracted
  from `https://curl.se/ca/cacert.pem`. Besides

>   the OpenSSL DLLs (e.g. libssl-1_1-x64.dll, libcrypto-1_1-x64.dll) you

Reviewer 8: ```suggestion
  the OpenSSL DLLs (e.g. `libssl-1_1-x64.dll`, `libcrypto-1_1-x64.dll`) you
```

  now also need to ship `cacert.pem` with your `.exe` file.

## `std/hashes`
- `hashes.hash` can now support `object` and `ref` (can be overloaded in user code),
  if `-d:nimEnableHashRef` is used.

> 

Reviewer 8: ```suggestion
```

## OS: `std/os`, `std/io`, `std/socketstream`, `std/linenoise`, `std/tempfiles`
- `os.FileInfo` (returned by `getFileInfo`) now contains `blockSize`,
  determining preferred I/O block size for this file object.
- Added `os.getCacheDir()` to return platform specific cache directory.
- Improved `os.getTempDir()`, see PR [#16914](https://github.com/nim-lang/Nim/pull/16914).
- Added `os.isAdmin` to tell whether the caller's process is a member of the
  Administrators local group (on Windows) or a root (on POSIX).
- Added optional `options` argument to `copyFile`, `copyFileToDir`, and
  `copyFileWithPermissions`. By default, on non-Windows OSes, symlinks are
  followed (copy files symlinks point to); on Windows, `options` argument is
  ignored and symlinks are skipped.
- On non-Windows OSes, `copyDir` and `copyDirWithPermissions` copy symlinks as
  symlinks (instead of skipping them as it was before); on Windows symlinks are
  skipped.
- On non-Windows OSes, `moveFile` and `moveDir` move symlinks as symlinks
  (instead of skipping them sometimes as it was before).
- Added optional `followSymlinks` argument to `setFilePermissions`.
- Added a simpler to use `io.readChars` overload.

> - Added `socketstream` module that wraps sockets in the stream interface

Reviewer 8: ```suggestion
- Added `socketstream` module that wraps sockets in the stream interface.
```

- Added experimental `linenoise.readLineStatus` to get line and status (e.g. ctrl-D or ctrl-C).

## Environment variable handling
- empty environment variable values are now supported across OS's and backends
- environment variable APIs now work in multithreaded scenarios, by delegating to direct OS calls
  instead of trying to keep track of the environment.
- `putEnv`, `delEnv` now work at CT.
- NodeJS backend now supports osenv: `getEnv`, `putEnv`, `envPairs`, `delEnv`, `existsEnv`.

Compatibility notes:

> - `std/os`: `putEnv` now raises if the 1st argument contains a `=`

Reviewer 8: ```suggestion
- `std/os`: `putEnv` now raises if the 1st argument contains a `=`.
```

## POSIX
- On POSIX systems, the default signal handlers used for Nim programs (it's
  used for printing the stacktrace on fatal signals) will now re-raise the
  signal for the OS default handlers to handle.
  This lets the OS perform its default actions, which might include core
  dumping (on select signals) and notifying the parent process about the cause
  of termination.
- On POSIX systems, we now ignore `SIGPIPE` signals, use `-d:nimLegacySigpipeHandler`
  for previous behavior.
- Added `posix_utils.osReleaseFile` to get system identification from `os-release` file on Linux and the BSDs.
  https://www.freedesktop.org/software/systemd/man/os-release.html

> - Remove undefined behavior for `posix.open`

Reviewer 8: ```suggestion
- Remove undefined behavior for `posix.open`.
```

## `std/prelude`
- `std/strformat` is now part of `include std/prelude`.
- Added `std/sequtils` import to `std/prelude`.
- `std/prelude` now works with the JS target.
- `std/prelude` can now be used via `include std/prelude`, but `include prelude` still works.

## String manipulation: `std/strformat`, `std/strbasics`

> - added support for parenthesized expressions.
> - added support for const string's instead of just string literals
> 
> - Added `std/strbasics` for high performance string operations.
>   Added `strip`, `setSlice`, `add(a: var string, b: openArray[char])`.

Reviewer 8: ```suggestion
- Added support for parenthesized expressions.
- Added support for const string's instead of just string literals

- Added `std/strbasics` for high performance string operations.
- Added `strip`, `setSlice`, `add(a: var string, b: openArray[char])`.
```

## `std/wrapnils`
- `std/wrapnils` doesn't use `experimental:dotOperators` anymore, avoiding
  issues like bug [#13063](https://github.com/nim-lang/Nim/issues/13063) (which affected error messages)
  for modules importing `std/wrapnils`.

>   Added `??.` macro which returns an `Option`.
>   `std/wrapnils` can now be used to protect against `FieldDefect` errors in

Reviewer 8: ```suggestion
- Added `??.` macro which returns an `Option`.
- `std/wrapnils` can now be used to protect against `FieldDefect` errors in
```

  case objects, generates optimal code (no overhead compared to manual
  if-else branches), and preserves lvalue semantics which allows modifying
  an expression.

## Containers: `std/algorithm`, `std/lists`, `std/sequtils`, `std/options`, `std/packedsets`
- Removed the optional `longestMatch` parameter of the `critbits._WithPrefix` iterators (it never worked reliably)
- Added `algorithm.merge`.
- In `std/lists`: renamed `append` to `add` and retained `append` as an alias;
  added `prepend` and `prependMoved` analogously to `add` and `addMoved`;
  added `remove` for `SinglyLinkedList`s.
- Added new operations for singly- and doubly linked lists: `lists.toSinglyLinkedList`
  and `lists.toDoublyLinkedList` convert from `openArray`s; `lists.copy` implements
  shallow copying; `lists.add` concatenates two lists - an O(1) variation that consumes
  its argument, `addMoved`, is also supplied. 
  See PR [#16362](https://github.com/nim-lang/Nim/pull/16362), [#16536](https://github.com/nim-lang/Nim/pull/16536).

> - new module: `std/packedsets`

Reviewer 8: ```suggestion
- New module: `std/packedsets`
```

  Generalizes `std/intsets`, see PR [#15564](https://github.com/nim-lang/Nim/pull/15564).

Compatibility notes:
- Deprecated `sequtils.delete` and added an overload taking a `Slice` that raises a defect
  if the slice is out of bounds, likewise with `strutils.delete`.
- Deprecated `proc reversed*[T](a: openArray[T], first: Natural, last: int): seq[T]` in `std/algorithm`.
- `std/options` changed `$some(3)` to `"some(3)"` instead of `"Some(3)"`
  and `$none(int)` to `"none(int)"` instead of `"None[int]"`.

## `std/times`
- Added `ZZZ` and `ZZZZ` patterns to `times.nim` `DateTime` parsing, to match time
  zone offsets without colons, e.g. `UTC+7 -> +0700`.
- Added `dateTime` and deprecated `initDateTime`.

## `std/macros` and AST
- New module `std/genasts` containing `genAst` that avoids the problems inherent with `quote do` and can
  be used as a replacement.

>   use `-d:nimLegacyMacrosCollapseSymChoice` to get previous behavior.

Reviewer 8: ```suggestion
  Use `-d:nimLegacyMacrosCollapseSymChoice` to get the previous behavior.
```

- `strscans.scanTuple` added which uses `strscans.scanf` internally,
  returning a tuple which can be unpacked for easier usage of `scanf`.
- Added `decodeQuery` to `std/uri`.
- `parseopt.initOptParser` has been made available and `parseopt` has been
  added back to `std/prelude` for all backends. Previously `initOptParser` was
  unavailable if the `std/os` module did not have `paramCount` or `paramStr`,
  but the use of these in `initOptParser` were conditionally to the runtime
  arguments passed to it, so `initOptParser` has been changed to raise
  `ValueError` when the real command line is not available. `parseopt` was
  previously excluded from `std/prelude` for JS, as it could not be imported.

Compatibility notes:
- Changed the behavior of `uri.decodeQuery` when there are unencoded `=`
  characters in the decoded values. Prior versions would raise an error. This is
  no longer the case to comply with the HTML spec and other languages
  implementations. Old behavior can be obtained with
  `-d:nimLegacyParseQueryStrict`. `cgi.decodeData` which uses the same
  underlying code is also updated the same way.

## JS stdlib changes

> - Added `std/jsbigints` module, arbitrary precision integers for JS target.
> - Added `setCurrentException` for JS backend.
> - `writeStackTrace` is available in JS backend now.
> - Added `then`, `catch` to `std/asyncjs` for promise pipelining, for now hidden behind `-d:nimExperimentalAsyncjsThen`.
> - Added `std/jsfetch` module [Fetch](https://developer.mozilla.org/docs/Web/API/Fetch_API) wrapper for JS target.
> - Added `std/jsheaders` module [Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers) wrapper for JS target.
> - Added `std/jsformdata` module [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) wrapper for JS target.
> - Added `jscore.debugger` to [call any available debugging functionality, such as breakpoints.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/debugger).
> - Added `jsconsole.dir`, `jsconsole.dirxml`, `jsconsole.timeStamp`.
> - Added dollar `$` and `len` for `jsre.RegExp`.
> - Added `jsconsole.jsAssert` for JS target.
> - Added `**` to `std/jsffi`.
> - Added `copyWithin` [for `seq` and `array` for JS targets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/copyWithin).
> - In `std/dom`, `Interval` is now a `ref object`, same as `Timeout`. Definitions of `setTimeout`,
>   `clearTimeout`, `setInterval`, `clearInterval` were updated.
> - Added `dom.scrollIntoView` proc with options
> - Added `dom.setInterval`, `dom.clearInterval` overloads.
> - Merged `std/dom_extensions` module into `std/dom` module,
>   it was a module with a single line, see RFC [#413](https://github.com/nim-lang/RFCs/issues/413).
> - `$` now gives more correct results on JS backend.

Reviewer 8: ```suggestion
- Added `std/jsbigints` module, arbitrary precision integers for the JS target.
- Added `setCurrentException` for the JS backend.
- `writeStackTrace` is available in the JS backend now.
- Added `then`, `catch` to `std/asyncjs` for promise pipelining, for now hidden behind `-d:nimExperimentalAsyncjsThen`.
- Added `std/jsfetch` module [Fetch](https://developer.mozilla.org/docs/Web/API/Fetch_API) wrapper for the JS target.
- Added `std/jsheaders` module [Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers) wrapper for the JS target.
- Added `std/jsformdata` module [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) wrapper for the JS target.
- Added `jscore.debugger` to [call any available debugging functionality, such as breakpoints](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/debugger).
- Added `jsconsole.dir`, `jsconsole.dirxml`, `jsconsole.timeStamp`.
- Added dollar `$` and `len` for `jsre.RegExp`.
- Added `jsconsole.jsAssert` for the JS target.
- Added `**` to `std/jsffi`.
- Added `copyWithin` [for `seq` and `array` for JS targets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/copyWithin).
- In `std/dom`, `Interval` is now a `ref object`, same as `Timeout`. Definitions of `setTimeout`,
  `clearTimeout`, `setInterval`, `clearInterval` were updated.
- Added `dom.scrollIntoView` proc with options.
- Added `dom.setInterval`, `dom.clearInterval` overloads.
- Merged `std/dom_extensions` into the `std/dom` module,
  it was a module with a single line, see RFC [#413](https://github.com/nim-lang/RFCs/issues/413).
- `$` now gives more correct results on the JS backend.
```

## JS compiler changes

> - The `cstring` doesn't support `[]=` operator in JS backend.
> - Now array literals(JS backend) uses JS typed arrays when the corresponding JS typed array exists,
>   for example `[byte(1), 2, 3]` generates `new Uint8Array([1, 2, 3])`.

Reviewer 8: ```suggestion
- `cstring` doesn't support the `[]=` operator anymore in the JS backend.
- Array literals now use JS typed arrays when the corresponding JS typed array exists,
  for example `[byte(1), 2, 3]` generates `new Uint8Array([1, 2, 3])`.
```

## VM and nimscript backend

> - VM now supports `addr(mystring[ind])` (index + index assignment)
> - `nimscript` now handles `except Exception as e`.
> - nil dereference is not allowed at compile time. `cast[ptr int](nil)[]` is rejected at compile time.
> - `static[T]` now works better, refs [#17590](https://github.com/nim-lang/Nim/pull/17590), [#15853](https://github.com/nim-lang/Nim/pull/15853).
> - `distinct T` conversions now works in VM.
> - `items(cstring)` now works in VM
> - fix `addr`, `len`, `high` in VM ([#16002](https://github.com/nim-lang/Nim/pull/16002), [#16610](https://github.com/nim-lang/Nim/pull/16610)).

Reviewer 8: ```suggestion
- VM now supports `addr(mystring[ind])` (index + index assignment).
- `nimscript` now handles `except Exception as e`.
- `nil` dereference is not allowed at compile time. `cast[ptr int](nil)[]` is rejected at compile time.
- `static[T]` now works better, refs [#17590](https://github.com/nim-lang/Nim/pull/17590), [#15853](https://github.com/nim-lang/Nim/pull/15853).
- `distinct T` conversions now work in VM.
- `items(cstring)` now works in VM.
- Fix `addr`, `len`, `high` in VM ([#16002](https://github.com/nim-lang/Nim/pull/16002), [#16610](https://github.com/nim-lang/Nim/pull/16610)).
```

- `std/cstrutils` now works in VM.

## OS-specific notes
- Support for Apple silicon/M1.
- Support for 32-bit RISC-V, refs [#16231](https://github.com/nim-lang/Nim/pull/16231).
- Support for CROSSOS, refs [#18889](https://github.com/nim-lang/Nim/pull/18889).
- The allocator for Nintendo Switch, which was nonfunctional because
  of breaking changes in libnx, was removed, in favour of the new `-d:nimAllocPagesViaMalloc` option.
- Allow reading parameters when compiling for Nintendo Switch.

- Cross compilation targeting Windows was improved

>   This now works from OSX/Linux:

Reviewer 8: ```suggestion
  This now works from macOS/Linux:
```

  `nim r -d:mingw main`
  `--nimcache` now correctly works in a cross-compilation setting.

## Performance / memory optimizations
- The comment field in PNode AST was moved to a side channel, reducing overall memory usage during compilation by a factor 1.25x

> - `std/jsonutils` deserialization is now up to 20x faster

Reviewer 8: ```suggestion
- `std/jsonutils` deserialization is now up to 20x faster.
```

> - `os.copyFile` is now 2.5x faster on OSX, by using `copyfile` from `copyfile.h`;
>   use `-d:nimLegacyCopyFile` for OSX < 10.5.

Reviewer 8: ```suggestion
- `os.copyFile` is now 2.5x faster on macOS, by using `copyfile` from `copyfile.h`;
  use `-d:nimLegacyCopyFile` for macOS < 10.5.
```

> - float to string is now 10x faster thanks to Dragonbox algorithm, with `-d:nimPreviewFloatRoundtrip`.
> - `newSeqWith` is 3x faster
> - CI now supports batching (making Windows CI 2.3X faster).
> - sets now uses optimized `countSetBits`, see PR [#17334](https://github.com/nim-lang/Nim/pull/17334).

Reviewer 8: ```suggestion
- Float to string conversion is now 10x faster thanks to the Dragonbox algorithm, with `-d:nimPreviewFloatRoundtrip`.
- `newSeqWith` is 3x faster.
- CI now supports batching (making Windows CI 2.3X faster).
- Sets now uses the optimized `countSetBits` proc, see PR [#17334](https://github.com/nim-lang/Nim/pull/17334).
```

  This warning will become an error in future versions! Use an explicit conversion
  like `cstring(x)` in order to silence the warning.

- There is a new warning for *any* type conversion to `enum` that can be enabled via
  `.warning[AnyEnumConv]:on` or `--warning:AnyEnumConv:on`.

- Reusing a type name in a different scope now works, refs [#17710](https://github.com/nim-lang/Nim/pull/17710).
- Fixed implicit and explicit generics in procedures, refs [#18808](https://github.com/nim-lang/Nim/pull/18808).

## New-style concepts
Example:
```nim
type
  Comparable = concept # no T, an atom
    proc cmp(a, b: Self): int
```
See PR [#15251](https://github.com/nim-lang/Nim/pull/15251) for details.

## Lexical / syntactic
- Nim now supports a small subset of Unicode operators as operator symbols.
  The supported symbols are: "∙ ∘ × ★ ⊗ ⊘ ⊙ ⊛ ⊠ ⊡ ∩ ∧ ⊓ ± ⊕ ⊖ ⊞ ⊟ ∪ ∨ ⊔".
  To enable this feature, use `--experimental:unicodeOperators`. Note that due
  to parser limitations you **cannot** enable this feature via a
  pragma `{.experimental: "unicodeOperators".}` reliably, you need to enable
  it via the command line or in a configuration file.

> - `var a{.foo.} = expr` now works inside templates (except when `foo` is overloaded).

Reviewer 8: ```suggestion
- `var a {.foo.} = expr` now works inside templates (except when `foo` is overloaded).
```

- Added `--hintAsError` with similar semantics as `--warningAsError`.
- Added `--unitsep:on|off` to control whether to add ASCII unit separator `\31` before a newline
 for every generated message (potentially multiline), so tooling can tell when messages start and end.
- Added `--filenames:abs|canonical|legacyRelProj` which replaces `--listFullPaths:on|off`
- `--hint:all:on|off` is now supported to select or deselect all hints; it
  differs from `--hints:on|off` which acts as a (reversible) gate.
  Likewise with `--warning:all:on|off`.
- The style checking of the compiler now supports a `--styleCheck:usages` switch. This switch
  enforces that every symbol is written as it was declared, not enforcing
  the official Nim style guide. To be enabled, this has to be combined either
  with `--styleCheck:error` or `--styleCheck:hint`.
- Type mismatch errors now show more context, use `-d:nimLegacyTypeMismatch` for previous
  behavior.
- `typedesc[Foo]` now renders as such instead of `type Foo` in compiler messages.
- `runnableExamples` now show originating location in stacktraces on failure.
- `SuccessX` message now shows more useful information.

> - new `DuplicateModuleImport` warning, and improved `UnusedImport` and `XDeclaredButNotUsed` accuracy.

Reviewer 8: ```suggestion
- New `DuplicateModuleImport` warning, improved `UnusedImport` and `XDeclaredButNotUsed` accuracy.
```

Compatibility notes:

> - `--hint:CC` now goes to stderr (like all other hints) instead of stdout.

Reviewer 8: ```suggestion
- `--hint:CC` now prints to stderr (like all other hints) instead of stdout.
```

  compiled with different options if the output file differs.

- `--usenimcache` (implied by `nim r main`) now generates an output file that includes a hash of
  some of the compilation options, which allows caching generated binaries:
```bash
  nim r main # recompiles
  nim r -d:foo main # recompiles
  nim r main # uses cached binary
  nim r main arg1 arg2 # ditto (runtime arguments are irrelevant)
```

- `nim r` now supports cross compilation from unix to windows when specifying `-d:mingw` by using Wine,
  e.g.: `nim r --eval:'import os; echo "a" / "b"'` prints `a\b`

- `nim` can compile version 1.4.0 as follows: `nim c --lib:lib --stylecheck:off -d:nimVersion140 compiler/nim`.
  `-d:nimVersion140` is not needed for bootstrapping, only for building 1.4.0 from devel.

- `nim e` now accepts arbitrary file extensions for the nimscript file,
  although `.nims` is still the preferred extension in general.
- The configuration subsystem now allows for `-d:release` and `-d:danger` to work as expected.
  The downside is that these defines now have custom logic that doesn't apply for
  other defines.

## Multithreading

> - TLS: OSX now uses native TLS (`--tlsEmulation:off`), TLS now works with importcpp non-POD types,

Reviewer 8: ```suggestion
- TLS: macOS now uses native TLS (`--tlsEmulation:off`), TLS now works with `importcpp` non-POD types,
```

- Added `unsafeIsolate` and `extract` to `std/isolation`.
- Added `std/tasks`, a new module containing primitives for creating parallel programs.

## Memory management
- `--gc:arc` now bootstraps (PR [#17342](https://github.com/nim-lang/Nim/pull/17342)).
- Lots of improvements to `gc:arc`, `gc:orc`, refs PR [#15697](https://github.com/nim-lang/Nim/pull/15697), [#16849](https://github.com/nim-lang/Nim/pull/16849), [#17993](https://github.com/nim-lang/Nim/pull/17993).
- `--gc:orc` is now 10% faster than previously for common workloads. If
  you have trouble with its changed behavior, compile with `-d:nimOldOrc`.
- The `--gc:orc` algorithm was refined so that custom container types can participate in the
  cycle collection process. See the documentation of `=trace` for more details.
- On embedded devices `malloc` can now be used instead of `mmap` via `-d:nimAllocPagesViaMalloc`.
  This is only supported for `--gc:orc` or `--gc:arc`.

Compatibility notes:
- `--newruntime` and `--refchecks` are deprecated,
  use `--gc:arc`, `--gc:orc`, or `--gc:none` as appropriate instead.

## Docgen
- docgen: RST files can now use single backticks instead of double backticks and correctly render

>   in both `rst2html` (as before) as well as common tools rendering RST directly (e.g. GitHub), by
>   adding: `default-role:: code` directive inside the RST file, which is now handled by `rst2html`.

Reviewer 8: ```suggestion
  in both `rst2html` (as before) as well as common tools rendering RST directly (e.g. GitHub). This is done by
  adding the `default-role:: code` directive inside the RST file (which is now handled by `rst2html`).
```

  - every table row must start with `|`, e.g. `| cell 1 | cell 2 |`.

- Implemented `doc2tex` compiler command which converts documentation in
  `.nim` files to Latex.

- docgen now supports syntax highlighting for inline code.
- docgen now supports same line doc comments:
```nim
func fn*(a: int): int = 42  ## Doc comment
```
- docgen now renders deprecated and other pragmas.

- `runnableExamples` now works with templates and nested templates.
- `runnableExamples: "-r:off"` now works for examples that should compile but not run. 
- `runnableExamples` now renders code verbatim, and produces correct code in all cases.
- docgen now shows correct, canonical import paths in docs.
- docgen now shows all routines in sidebar, and the proc signature is now shown in sidebar.

## Effects and checks
- Significant improvement to error messages involving effect mismatches

- There is a new `cast` section `{.cast(uncheckedAssign).}: body` that disables some
  compiler checks regarding `case objects`. This allows serialization libraries
  to avoid ugly, non-portable solutions. See RFC [#407](https://github.com/nim-lang/RFCs/issues/407).
  for more details.

## Tools

> - major improvements to `nimgrep`, see PR [#15612

Reviewer 8: ```suggestion
- Major improvements to `nimgrep`, see PR [#15612
```

You can also use it to explore a module's APIs, including private symbols:
```bash
nim --eval:'import os {.all.}; echo weirdTarget'
```
See PR [#15687](https://github.com/nim-lang/Nim/pull/15687) for more details.

## Round-trip float to string
`system.addFloat` and `system.$` now can produce string representations of floating point numbers that are minimal in size and possess round-trip and correct rounding guarantees (via the [Dragonbox](https://raw.githubusercontent.com/jk-jeon/dragonbox/master/other_files/Dragonbox.pdf) algorithm).
This currently has to be enabled via `-d:nimPreviewFloatRoundtrip`.
It is expected that this behavior becomes the new default in upcoming versions, as with other `nimPreviewX` define flags.

Example:
```nim
from math import round
let a = round(9.779999999999999, 2)
assert a == 9.78
echo a # with `-d:nimPreviewFloatRoundtrip`: 9.78, like in python3 (instead of  9.779999999999999)
```

## New `std/jsbigints` module
Provides arbitrary precision integers for the JS target. See PR [#16409](https://github.com/nim-lang/Nim/pull/16409).
Example:
```nim
import std/jsbigints
assert 2'big ** 65'big == 36893488147419103232'big

> echo 0xdeadbeaf'big shl 4'big # 59774855920n

Reviewer 8: ```suggestion
echo 0xdeadbeef'big shl 4'big # 59774856944
```

Author: the `n` indicates bigints in js, see for yourself :)
```
nim r -b:js --eval:"import std/jsbigints; echo 0xdeadbeaf'big shl 4'big"
```

Reviewer 8: Ah, it looked like a typo. The other change was `0xdeadbeaf` to `0xdeadbeef`, the latter is what people actually use.

Author: oh ya, missed that part, fixed

---
title: "Version 1.6.0 released"
author: The Nim Team
---

Nim version 1.6 is now officially released!

A year in the making, 1.6 is the latest stable release and by far the largest yet.
We're proud of what we, the core team and dedicated volunteers, have accomplished with this milestone.
Here are some stats:
* 1667 PRs merged (1760 commits)
* 893 issues closed
* 15 new stdlib modules

> * new features in >40 stdlib modules, including major improvements to 10 commonly used modules.

Reviewer 6: ```suggestion
* new features in more than 40 stdlib modules, including major improvements to 10 commonly used modules.
```

# Major new features
With so many new features, pinpointing the most salient ones is a subjective exercise,
but here are a select few:

## Strict effects
The effect system was refined and there is a new `.effectsOf` annotation that does
explicitly what was previously done implicitly. See the [manual](https://nim-lang.github.io/Nim/manual.html#effect-system-effectsof-annotation) for more details.
To write code that is portable with older Nim versions, use this idiom:

```nim
when defined(nimHasEffectsOf):
  {.experimental: "strictEffects".}
else:
  {.pragma: effectsOf.}

proc mysort(s: seq; cmp: proc(a, b: T): int) {.effectsOf: cmp.}
```

To enable the new effect system, compile with `--experimental:strictEffects`. See also [#18777](https://github.com/nim-lang/Nim/pull/18777) and RFC [#408](https://github.com/nim-lang/RFCs/issues/408).

## `iterable[T]`
The `iterable[T]` type class was added to match called iterators, which enables writing:
```nim
template fn(a: iterable) # or template fn[T](a: iterable[T])
# instead of:
template fn(a: untyped)
```
This solves a number of long standing issues related to iterators. In particular,

> iterable arguments can now be used with MCS, e.g. `iota(3).toSeq` now works.

Author: I've now included a complete self-contained example which illustrates the main points about iterable, with a `sum` example showing before/after

This implies that edge cases like `-128'i8` finally work correctly.
Example:
```nim
func `'big`*(num: cstring): JsBigInt {.importjs: "BigInt(#)".}
assert 0xffffffffffffffff'big == (1'big shl 64'big) - 1'big
```

## Dot-like operators
With `-d:nimPreviewDotLikeOps`, dot-like operators (operators starting with `.`, but not with `..`)
now have the same precedence as `.`, so that `a.?b.c` is now parsed as `(a.?b).c` instead of `a.?(b.c)`.
A warning is generated when a dot-like operator is used without `-d:nimPreviewDotLikeOps`.

An important use case is to enable dynamic fields without affecting the builtin `.` operator, e.g. for
`std/jsffi`, `std/json`, `pkg/nimpy`. Example:
```nim
import std/json
template `.?`(a: JsonNode, b: untyped{ident}): JsonNode =
  a[astToStr(b)]
let j = %*{"a1": {"a2": 10}}
assert j.?a1.?a2.getInt == 10
```

## Block arguments now support optional parameters
This solves a major pain point for routines accepting block parameters, see PR #18631 for details:
```nim
template fn(a = 1, b = 2, body) = discard
fn(1, 2): # already works
  bar
fn(a = 1): # now works
  bar
```

> ditto with multiple block arguments via `do`:

Author: i've now changed all ditto instances to likewise

- Removed deprecated `std/iup` module from stdlib, it has already moved to
  [nimble](https://github.com/nim-lang/iup).

## New module: `std/setutils`
- Added `setutils.toSet` that can take any iterable and convert it to a built-in `set`,
  if the iterable yields a built-in settable type.
- Added `setutils.fullSet` which returns a full built-in `set` for a valid type.
- Added `setutils.complement` which returns the complement of a built-in `set`.
- Added `setutils.[]=`.

## New module: `std/enumutils`
- Added `genEnumCaseStmt` macro that generates
  case statement to parse string to enum.
- Added `items` for enums with holes.
- Added `symbolName` to return the `enum` symbol name ignoring the human readable name.
- Added `symbolRank` to return the index in which an `enum` member is listed in an enum.

## `system`
- Added `system.prepareMutation` for better support of low
  level `moveMem`, `copyMem` operations for `gc:orc`'s copy-on-write string
  implementation.
- `system.addEscapedChar` now renders `\r` as `\r` instead of `\c`, to be compatible
  with most other languages.
- Added `cmpMem` to `system`.
- `doAssertRaises` now correctly handles foreign exceptions.

> - `addInt` now supports unsigned integers

Reviewer 9: ```suggestion
- `addInt` now supports unsigned integers.
```

## Environment variable handling
- empty environment variable values are now supported across OS's and backends
- environment variable APIs now work in multithreaded scenarios, by delegating to direct OS calls
  instead of trying to keep track of the environment.
- `putEnv`, `delEnv` now work at CT.
- NodeJS backend now supports osenv: `getEnv`, `putEnv`, `envPairs`, `delEnv`, `existsEnv`.

Compatibility notes:
- `std/os`: `putEnv` now raises if the 1st argument contains a `=`.

## POSIX
- On POSIX systems, the default signal handlers used for Nim programs (it's
  used for printing the stacktrace on fatal signals) will now re-raise the
  signal for the OS default handlers to handle.
  This lets the OS perform its default actions, which might include core
  dumping (on select signals) and notifying the parent process about the cause
  of termination.
- On POSIX systems, we now ignore `SIGPIPE` signals, use `-d:nimLegacySigpipeHandler`
  for previous behavior.
- Added `posix_utils.osReleaseFile` to get system identification from `os-release` file on Linux and the BSDs.
  https://www.freedesktop.org/software/systemd/man/os-release.html

> - Remove undefined behavior for `posix.open`.

Reviewer 6: ```suggestion
- Removed undefined behavior for `posix.open`.
```

## `std/prelude`
- `std/strformat` is now part of `include std/prelude`.
- Added `std/sequtils` import to `std/prelude`.
- `std/prelude` now works with the JS target.
- `std/prelude` can now be used via `include std/prelude`, but `include prelude` still works.

## String manipulation: `std/strformat`, `std/strbasics`
- Added support for parenthesized expressions.

> - Added support for const string's instead of just string literals

Reviewer 6: ```suggestion
- Added support for const string's instead of just string literals.
```

> 

Reviewer 6: ```suggestion
```

- Added `std/strbasics` for high performance string operations.
- Added `strip`, `setSlice`, `add(a: var string, b: openArray[char])`.

## `std/wrapnils`
- `std/wrapnils` doesn't use `experimental:dotOperators` anymore, avoiding
  issues like bug [#13063](https://github.com/nim-lang/Nim/issues/13063) (which affected error messages)
  for modules importing `std/wrapnils`.
- Added `??.` macro which returns an `Option`.
- `std/wrapnils` can now be used to protect against `FieldDefect` errors in
  case objects, generates optimal code (no overhead compared to manual
  if-else branches), and preserves lvalue semantics which allows modifying
  an expression.

## Containers: `std/algorithm`, `std/lists`, `std/sequtils`, `std/options`, `std/packedsets`

> - Removed the optional `longestMatch` parameter of the `critbits._WithPrefix` iterators (it never worked reliably)

Reviewer 6: ```suggestion
- Removed the optional `longestMatch` parameter of the `critbits._WithPrefix` iterators (it never worked reliably).
```

- Added `algorithm.merge`.
- In `std/lists`: renamed `append` to `add` and retained `append` as an alias;
  added `prepend` and `prependMoved` analogously to `add` and `addMoved`;
  added `remove` for `SinglyLinkedList`s.
- Added new operations for singly- and doubly linked lists: `lists.toSinglyLinkedList`
  and `lists.toDoublyLinkedList` convert from `openArray`s; `lists.copy` implements
  shallow copying; `lists.add` concatenates two lists - an O(1) variation that consumes
  its argument, `addMoved`, is also supplied. 
  See PR [#16362](https://github.com/nim-lang/Nim/pull/16362), [#16536](https://github.com/nim-lang/Nim/pull/16536).

> 

Reviewer 6: ```suggestion
```

> - New module: `std/packedsets`

Reviewer 6: ```suggestion
- New module: `std/packedsets`.
```

  Generalizes `std/intsets`, see PR [#15564](https://github.com/nim-lang/Nim/pull/15564).

Compatibility notes:
- Deprecated `sequtils.delete` and added an overload taking a `Slice` that raises a defect
  if the slice is out of bounds, likewise with `strutils.delete`.
- Deprecated `proc reversed*[T](a: openArray[T], first: Natural, last: int): seq[T]` in `std/algorithm`.
- `std/options` changed `$some(3)` to `"some(3)"` instead of `"Some(3)"`
  and `$none(int)` to `"none(int)"` instead of `"None[int]"`.

## `std/times`
- Added `ZZZ` and `ZZZZ` patterns to `times.nim` `DateTime` parsing, to match time
  zone offsets without colons, e.g. `UTC+7 -> +0700`.
- Added `dateTime` and deprecated `initDateTime`.

## `std/macros` and AST
- New module `std/genasts` containing `genAst` that avoids the problems inherent with `quote do` and can
  be used as a replacement.
  Use `-d:nimLegacyMacrosCollapseSymChoice` to get the previous behavior.

> 

Reviewer 6: ```suggestion
```

- The required name of case statement macros for the experimental
  `caseStmtMacros` feature has changed from `match` to `` `case` ``.
- Tuple expressions are now parsed consistently as
  `nnkTupleConstr` node. Will affect macros expecting nodes to be of `nnkPar`.

> - In `std/macros`, `treeRepr,lispRepr,astGenRepr` now represent SymChoice nodes in a collapsed way,

Reviewer 6: ```suggestion
- In `std/macros`, `treeRepr,lispRepr,astGenRepr` now represent SymChoice nodes in a collapsed way.
```

> 

Reviewer 6: ```suggestion
```

- Make custom op in `macros.quote` work for all statements.

## `std/sugar`
- Added `sugar.dumpToString` which improves on `sugar.dump`.

> - Added an overload for the `collect` macro that inferes the container type based

Reviewer 6: ```suggestion
- Added an overload for the `collect` macro that infers the container type based
```

  for example `[byte(1), 2, 3]` generates `new Uint8Array([1, 2, 3])`.

## VM and nimscript backend
- VM now supports `addr(mystring[ind])` (index + index assignment).
- `nimscript` now handles `except Exception as e`.
- `nil` dereference is not allowed at compile time. `cast[ptr int](nil)[]` is rejected at compile time.
- `static[T]` now works better, refs [#17590](https://github.com/nim-lang/Nim/pull/17590), [#15853](https://github.com/nim-lang/Nim/pull/15853).
- `distinct T` conversions now work in VM.
- `items(cstring)` now works in VM.
- Fix `addr`, `len`, `high` in VM ([#16002](https://github.com/nim-lang/Nim/pull/16002), [#16610](https://github.com/nim-lang/Nim/pull/16610)).
- `std/cstrutils` now works in VM.

## OS-specific notes
- Support for Apple silicon/M1.
- Support for 32-bit RISC-V, refs [#16231](https://github.com/nim-lang/Nim/pull/16231).
- Support for CROSSOS, refs [#18889](https://github.com/nim-lang/Nim/pull/18889).
- The allocator for Nintendo Switch, which was nonfunctional because
  of breaking changes in libnx, was removed, in favour of the new `-d:nimAllocPagesViaMalloc` option.
- Allow reading parameters when compiling for Nintendo Switch.

> 

Reviewer 6: ```suggestion
```

- `std/jsonutils` deserialization is now up to 20x faster.
- `os.copyFile` is now 2.5x faster on macOS, by using `copyfile` from `copyfile.h`;
  use `-d:nimLegacyCopyFile` for macOS < 10.5.
- Float to string conversion is now 10x faster thanks to the Dragonbox algorithm, with `-d:nimPreviewFloatRoundtrip`.
- `newSeqWith` is 3x faster.
- CI now supports batching (making Windows CI 2.3X faster).
- Sets now uses the optimized `countSetBits` proc, see PR [#17334](https://github.com/nim-lang/Nim/pull/17334).

## Debugging
- You can now enable/disable VM tracing in user code via `vmutils.vmTrace`.
- `koch tools` now builds `bin/nim_dbg` which allows easy access to a debug version of Nim without recompiling.
- Added new module `compiler/debugutils` to help with debugging Nim compiler.
- Renamed `-d:nimCompilerStackraceHints` to `-d:nimCompilerStacktraceHints` and used it in more contexts;
  this flag which works in tandem with `--stackTraceMsgs` to show user code context in compiler stacktraces.

## Type system
- `typeof(voidStmt)` now works and returns `void`.

> - `enum` values can now be overloaded. This needs to be enabled

Reviewer 6: ```suggestion

- `enum` values can now be overloaded. This needs to be enabled
```

Author: made vertical spacing consistent

  via `{.experimental: "overloadableEnums".}`. We hope that this feature allows for the
  development of more fluent (less ugly) APIs. See RFC [#373](https://github.com/nim-lang/RFCs/issues/373).
  for more details.

- A type conversion from one `enum` type to another now produces an `[EnumConv]` warning.
  You should use `ord` (or `cast`, but the compiler won't help, if you misuse it) instead.
  ```
  type A = enum a1, a2
  type B = enum b1, b2
  echo a1.B # produces a warning
  echo a1.ord.B # produces no warning
  ```

- A dangerous implicit conversion to `cstring` now triggers a `[CStringConv]` warning.
  This warning will become an error in future versions! Use an explicit conversion
  like `cstring(x)` in order to silence the warning.

- There is a new warning for *any* type conversion to `enum` that can be enabled via
  `.warning[AnyEnumConv]:on` or `--warning:AnyEnumConv:on`.

- Reusing a type name in a different scope now works, refs [#17710](https://github.com/nim-lang/Nim/pull/17710).

> - Fixed implicit and explicit generics in procedures, refs [#18808](https://github.com/nim-lang/Nim/pull/18808).

Reviewer 6: ```suggestion

- Fixed implicit and explicit generics in procedures, refs [#18808](https://github.com/nim-lang/Nim/pull/18808).
```

  behavior.
- `typedesc[Foo]` now renders as such instead of `type Foo` in compiler messages.
- `runnableExamples` now show originating location in stacktraces on failure.
- `SuccessX` message now shows more useful information.
- New `DuplicateModuleImport` warning, improved `UnusedImport` and `XDeclaredButNotUsed` accuracy.

Compatibility notes:
- `--hint:CC` now prints to stderr (like all other hints) instead of stdout.

## Building and running Nim programs, configuration system
- JSON build instructions are now generated in `$nimcache/outFileBasename.json`
  instead of `$nimcache/projectName.json`. This allows avoiding recompiling a given project
  compiled with different options if the output file differs.

- `--usenimcache` (implied by `nim r main`) now generates an output file that includes a hash of
  some of the compilation options, which allows caching generated binaries:
```bash
  nim r main # recompiles
  nim r -d:foo main # recompiles
  nim r main # uses cached binary
  nim r main arg1 arg2 # ditto (runtime arguments are irrelevant)
```

- `nim r` now supports cross compilation from unix to windows when specifying `-d:mingw` by using Wine,

>   e.g.: `nim r --eval:'import os; echo "a" / "b"'` prints `a\b`

Reviewer 6: ```suggestion
  e.g.: `nim r --eval:'import os; echo "a" / "b"'` prints `a\b`.
```

- `nim` can compile version 1.4.0 as follows: `nim c --lib:lib --stylecheck:off -d:nimVersion140 compiler/nim`.
  `-d:nimVersion140` is not needed for bootstrapping, only for building 1.4.0 from devel.

- `nim e` now accepts arbitrary file extensions for the nimscript file,
  although `.nims` is still the preferred extension in general.

> - The configuration subsystem now allows for `-d:release` and `-d:danger` to work as expected.

Reviewer 6: ```suggestion

- The configuration subsystem now allows for `-d:release` and `-d:danger` to work as expected.
```

Author: made vertical spacing consistent

- Lots of improvements to `gc:arc`, `gc:orc`, refs PR [#15697](https://github.com/nim-lang/Nim/pull/15697), [#16849](https://github.com/nim-lang/Nim/pull/16849), [#17993](https://github.com/nim-lang/Nim/pull/17993).
- `--gc:orc` is now 10% faster than previously for common workloads. If
  you have trouble with its changed behavior, compile with `-d:nimOldOrc`.
- The `--gc:orc` algorithm was refined so that custom container types can participate in the
  cycle collection process. See the documentation of `=trace` for more details.
- On embedded devices `malloc` can now be used instead of `mmap` via `-d:nimAllocPagesViaMalloc`.
  This is only supported for `--gc:orc` or `--gc:arc`.

Compatibility notes:
- `--newruntime` and `--refchecks` are deprecated,
  use `--gc:arc`, `--gc:orc`, or `--gc:none` as appropriate instead.

## Docgen
- docgen: RST files can now use single backticks instead of double backticks and correctly render
  in both `rst2html` (as before) as well as common tools rendering RST directly (e.g. GitHub). This is done by
  adding the `default-role:: code` directive inside the RST file (which is now handled by `rst2html`).

> 

Reviewer 6: ```suggestion
```

- Source+Edit links now appear on top of every docgen'd page when
  `nim doc --git.url:url ...` is given.

> 

Reviewer 6: ```suggestion
```

- Latex doc generation is revised: output `.tex` files should be compiled
  by `xelatex` (not by `pdflatex` as before). Now default Latex settings
  provide support for Unicode and do better job for avoiding margin overflows.

> 

Reviewer 6: ```suggestion
```

- The RST parser now supports footnotes, citations, admonitions, short style references with symbols.
- The RST parser now supports Markdown table syntax.
  Known limitations:
  - cell alignment is not supported, i.e. alignment annotations in a delimiter
    row (`:---`, `:--:`, `---:`) are ignored,
  - every table row must start with `|`, e.g. `| cell 1 | cell 2 |`.

> 

Reviewer 6: ```suggestion
```

- Implemented `doc2tex` compiler command which converts documentation in
  `.nim` files to Latex.

> 

Reviewer 6: ```suggestion
```

- docgen now supports syntax highlighting for inline code.
- docgen now supports same line doc comments:
```nim
func fn*(a: int): int = 42  ## Doc comment
```
- docgen now renders deprecated and other pragmas.

> 

Reviewer 6: ```suggestion
```

- `runnableExamples` now works with templates and nested templates.
- `runnableExamples: "-r:off"` now works for examples that should compile but not run. 
- `runnableExamples` now renders code verbatim, and produces correct code in all cases.
- docgen now shows correct, canonical import paths in docs.
- docgen now shows all routines in sidebar, and the proc signature is now shown in sidebar.

## Effects and checks
- Significant improvement to error messages involving effect mismatches

- There is a new `cast` section `{.cast(uncheckedAssign).}: body` that disables some
  compiler checks regarding `case objects`. This allows serialization libraries
  to avoid ugly, non-portable solutions. See RFC [#407](https://github.com/nim-lang/RFCs/issues/407).
  for more details.

## Tools
- Major improvements to `nimgrep`, see PR [#15612
](https://github.com/nim-lang/Nim/pull/15612).
- `fusion` is now un-bundled from Nim, `./koch fusion` will
  install it via Nimble at a fixed hash.

> 

Reviewer 6: ```suggestion
```

- `testament`: added `nimoutFull: bool` spec to compare full output of compiler
  instead of a subset; many bugfixes to testament.

## Misc/cleanups
- Deprecated `TaintedString` and `--taintmode`.
- Deprecated `--nilseqs` which is now a noop.

> - Added `-d:nimStrictMode` in CI in several places to ensure code doesn't have certain hints/warnings

Reviewer 6: ```suggestion
- Added `-d:nimStrictMode` in CI in several places to ensure code doesn't have certain hints/warnings.
```

- Removed `.travis.yml`, `appveyor.yml.disabled`, `.github/workflows/ci.yml.disabled`.

> - `[skip ci]` now works in azure and CI pipelines, see detail in PR [#17561](https://github.com/nim-lang/Nim/pull/17561)

Reviewer 6: ```suggestion
- `[skip ci]` now works in azure and CI pipelines, see detail in PR [#17561](https://github.com/nim-lang/Nim/pull/17561).
```

  [nimble](https://github.com/nim-lang/iup).

## New module: `std/setutils`
- Added `setutils.toSet` that can take any iterable and convert it to a built-in `set`,
  if the iterable yields a built-in settable type.
- Added `setutils.fullSet` which returns a full built-in `set` for a valid type.
- Added `setutils.complement` which returns the complement of a built-in `set`.
- Added `setutils.[]=`.

## New module: `std/enumutils`
- Added `genEnumCaseStmt` macro that generates
  case statement to parse string to enum.
- Added `items` for enums with holes.
- Added `symbolName` to return the `enum` symbol name ignoring the human readable name.
- Added `symbolRank` to return the index in which an `enum` member is listed in an enum.

## `system`
- Added `system.prepareMutation` for better support of low
  level `moveMem`, `copyMem` operations for `gc:orc`'s copy-on-write string
  implementation.
- `system.addEscapedChar` now renders `\r` as `\r` instead of `\c`, to be compatible
  with most other languages.
- Added `cmpMem` to `system`.
- `doAssertRaises` now correctly handles foreign exceptions.
- `addInt` now supports unsigned integers.

Compatibility notes:

> - `system.delete` had a most surprising behavior when the index passed to it was out of

Reviewer 3: ```suggestion
- `system.delete` had surprising behavior when the index passed to it was out of
```

Compatibility notes:
- `math.round` now rounds "away from zero" in the JS backend, which is consistent
  with other backends. See [#9125](https://github.com/nim-lang/Nim/pull/9125). Use `-d:nimLegacyJsRound` for the previous behavior.

## Random number generators: `std/random`, `std/sysrand`, `std/oids`
- Added `randState` template that exposes the default random number generator.
  Useful for library authors.
- Added `initRand()` overload with no argument which uses the current time as a seed.
- `initRand(seed)` now allows `seed == 0`.
- Added `std/sysrand` module to get random numbers from a secure source.
- Fixed overflow bugs.
- Fix `initRand` to avoid random number sequences overlapping, refs [#18744](https://github.com/nim-lang/Nim/pull/18744).
- `std/oids` now uses `std/random`.

Compatibility notes:
- Deprecated `std/mersenne`.
- `random.initRand(seed)` now produces non-skewed values for the 1st call to `rand()` after
  initialization with a small (< 30000) seed. Use `-d:nimLegacyRandomInitRand` to restore
  previous behavior for a transition time, see PR [#17467](https://github.com/nim-lang/Nim/pull/17467).

>   provided by the operating system.

Reviewer 3: ?

Author: fixed, see commit "improve note regarding std/sysrand"

---
title: "Version 1.6.0 released"
author: The Nim Team
---

Nim version 1.6 is now officially released!

A year in the making, 1.6 is the latest stable release and by far the largest yet.
We're proud of what we, the core team and dedicated volunteers, have accomplished with this milestone:
* 1667 PRs merged (1760 commits)
* 893 issues closed
* 15 new stdlib modules
* new features in more than 40 stdlib modules, including major improvements to 10 commonly used modules.
* documentation and minor improvements to 170 modules, including 312 new runnable examples
* 280 new nimble packages

Nim made its first entry in TIOBE index in 2017 at position 129,
last year it entered the top-100, and in the last 2 months the top-50 ([https://forum.nim-lang.org/t/8297](https://forum.nim-lang.org/t/8297)).
We hope this release will reinforce this trend, building on Nim's core strengths:
a practical, compiled systems programming language; offering C-like performance and portability;

> Python-like syntax; LISP-like flexibility; strong C, C++, JS, Python interop;

Reviewer 9: ```suggestion
Python-like syntax; Lisp-like flexibility; strong C, C++, JS, Python interop;
```

From wiki
> Lisp (historically LISP) is a family of programming languages with a long history and a distinctive, fully parenthesized prefix notation.

and best-in class metaprogramming.

Why use Nim?

> * one language to rule them all: write your iOS/android/web frontend and backend in the same language, from a [gameboy kickstarter](https://forum.nim-lang.org/t/8375) to a [blockchain client](https://github.com/status-im).

Reviewer 8: ```suggestion
* one language to rule them all: write your iOS/Android/web frontend and backend in the same language, from a [Gameboy kickstarter](https://forum.nim-lang.org/t/8375) to a [blockchain client](https://github.com/status-im).
```
Does Nim have a working Android backend?

Author: It's possible to use nim on android, although better support and adding android to CI would be things we'd need to do; eg:
* https://nim-lang.github.io/Nim/nimc.html#crossminuscompilation-for-android
* https://our.status.im/building-nimbus-on-android/
* https://forum.nim-lang.org/t/5197
and other links

> * one language to rule them all: write your iOS/android/web frontend and backend in the same language, from a [gameboy kickstarter](https://forum.nim-lang.org/t/8375) to a [blockchain client](https://github.com/status-im).

Reviewer 9: ```suggestion
* one language to rule them all: write your iOS/Android/web frontend and backend in the same language, from a [gameboy kickstarter](https://forum.nim-lang.org/t/8375) to a [blockchain client](https://github.com/status-im).
```

> * one language to rule them all: write your iOS/Android/web frontend and backend in the same language, from a [Gameboy kickstarter](https://forum.nim-lang.org/t/8375) to a [blockchain client](https://github.com/status-im).

Reviewer 10: This sentence is kind of awkward, our Kickstarter isn't an example of the same language being used for frontend and backend. The project is already mentioned in another bullet point so you don't have to mention it here (also in general I'd prefer not to be seen right next to blockchain stuff) :sweat_smile:

Author: sure; so we need to come up with something else in its place; suggestions from the community welcome

Reviewer 9: one language to rule them all: Good for everything, from [shell scripting](https://nim-lang.org/docs/nims.html) to [web front and backend](https://github.com/nim-lang/nimforum), to [scientific computing](https://github.com/SciNim), [blockchain client](https://github.com/status-im), [gamedev](https://github.com/ftsf/nico), and [embedded](https://github.com/EmbeddedNim).

Author: thanks, good, i edited this a bit and added `[deep learning](https://github.com/mratsim/Arraymancer)`

this line is important, we should polish it and find salient examples, further ideas / edits welcome

Reviewer 9: There are two games can be installed on android and iOS
https://www.frayhem.com/en/ (based on godot-nim)
https://github.com/pragmagic/godot-nim#made-with-godot-nim
https://oftengames.com/ (should be based on nimx or [rod](https://github.com/yglukhov/rod) I'm not sure.

Author: see last few edits, how does it look now?

> * concise and readable like python: `echo "hello world"` is a 1-liner (5 lines in C, C++, go, java).
> * small binaries: `echo "hello world"` generates a 73K binary (go: 2MB, rust: 377K, C++: 56K), thanks to dead code elimination [1].
> * fast compile times: a full compiler rebuild takes ~12s (rust: 15mn, gcc: 30mn+, clang: 1hr+, go: 90s) [2].

Reviewer 8: ```suggestion
* concise and readable like Python: `echo "hello world"` is a 1-liner (5 lines in C, C++, Go, Java).
* small binaries: `echo "hello world"` generates a 73K binary (Go: 2MB, Rust: 377K, C++: 56K) [1].
* fast compile times: a full compiler rebuild takes ~12s (Rust: 15min, gcc: 30min+, clang: 1hr+, Go: 90s) [2].
```
"thanks to dead code elimination" sounds like that's the only reason the Nim binaries are smaller, but I highly doubt that those other languages don't have dead code elimination).

Reviewer 6: ```suggestion
* concise and readable like python: `echo "hello world"` is a 1-liner (5 lines in C, C++, go, java).
* small binaries: `echo "hello world"` generates a 73K binary (go: 2MB, rust: 377K, C++: 56K), thanks to dead code elimination [1].
* fast compile times: a full compiler rebuild takes ~12s (rust: 15min, gcc: 30min+, clang: 1hr+, go: 90s) [2].
```

"The SI symbol for minute or minutes is **min**"
— https://en.wikipedia.org/wiki/Minute

* C-like performance: see [Web Frameworks Benchmark](https://web-frameworks-benchmark.netlify.app/result), [ray tracing](https://nim-lang.org/blog/2020/06/30/ray-tracing-in-nim.html), [primes](https://github.com/PlummersSoftwareLLC/Primes)
* no need for makefiles, cmake, configure or other build scripts, thanks to CTFE and dependency tracking [3]

> * target any platform with a C compiler: [embedded systems](https://github.com/elcritch/nesper), [micro-controllers](https://forum.nim-lang.org/t/7731), [WASM](https://forum.nim-lang.org/t/4779), Nintendo Switch or [Gameboy](https://www.kickstarter.com/projects/penguinrik/goodboy-galaxy-exploration-platform-game-gba-pc-and-switch)

Reviewer 10: This should be "Game Boy Advance", and can link to the forum post (https://forum.nim-lang.org/t/8375).

Note that the original Game Boy is way less C-friendly than GBA ^^

Author: thanks! done

Reviewer 3: Not sure if we should link the NLVM thread for wasm, as nlvm is quite outdated (Nim 1.2) compared to using Emscripten

Reviewer 9: Let's link nico (web game based on WASM) instead or treeform's tutorial instead!

> * target any platform with a C compiler: [embedded systems](https://github.com/elcritch/nesper), [micro-controllers](https://forum.nim-lang.org/t/7731), [WASM](https://forum.nim-lang.org/t/4779), nintendo switch or [gameboy](https://www.kickstarter.com/projects/penguinrik/goodboy-galaxy-exploration-platform-game-gba-pc-and-switch)

Reviewer 8: ```suggestion
* target any platform with a C compiler: [embedded systems](https://github.com/elcritch/nesper), [micro-controllers](https://forum.nim-lang.org/t/7731), [WASM](https://forum.nim-lang.org/t/4779), Nintendo Switch or [Gameboy](https://www.kickstarter.com/projects/penguinrik/goodboy-galaxy-exploration-platform-game-gba-pc-and-switch)
```

> * 0-overhead interop lets you reuse code in C, C++ (including templates, [C++ STL](https://clonkk.github.io/nim-cppstl/cppstl.html)), JS, Objective-C, python (via [nimpy](https://github.com/yglukhov/nimpy))

Reviewer 8: ```suggestion
* 0-overhead interop lets you reuse code in C, C++ (including templates, [C++ STL](https://clonkk.github.io/nim-cppstl/cppstl.html)), JS, Objective-C, Python (via [nimpy](https://github.com/yglukhov/nimpy))
```

> * built-in [documentation generator](https://nim-lang.github.io/Nim/system.html) that understands nim code and runnable examples that stay in sync

Reviewer 8: ```suggestion
* built-in [documentation generator](https://nim-lang.github.io/Nim/system.html) that understands Nim code and runnable examples that stay in sync
```

> Last but not least, `macros` let you manipulate/generate code at compile time instead of relying on code generators, allowing for example to easily write DSLs or extend the language, all in user code:

Reviewer 8: ```suggestion
Last but not least, macros let you manipulate/generate code at compile time instead of relying on code generators, allowing for example to easily write DSLs or extend the language, all in user code:
```
Using backticks implies to me that `macros` is some keyword or library (i know that `std/macros` exists, but that's not what's meant here).

  - [karax](https://github.com/karaxnim/karax) for react+JSX-like Single Page Applications
  - [cligen](https://github.com/c-blake/cligen) for API-inferred command line generator
  - [std/wrapnils](https://nim-lang.github.io/Nim/wrapnils.html) JS-like optional chaining

>   - [std/strformat](https://nim-lang.github.io/Nim/strformat.html) python-like f-strings

Reviewer 8: ```suggestion
  - [std/strformat](https://nim-lang.github.io/Nim/strformat.html) Python-like f-strings
```

>   - [std/strformat](https://nim-lang.github.io/Nim/strformat.html) Python-like f-strings

Reviewer: This is too much. This post should focus on the news, the blurb above is enough to entice those that don't know what Nim is about it. I would suggest replacing all of this with a link to your front page:

> Want to learn more about Nim and why you should consider using it? Take a look at a comprehensive list of its features [here][(https://nim-lang.org)

Reviewer: also it might be better to write a dedicated article for Goodboy (The Gameboy Advance game developed by @exelotl )

Author: I've shortened things a bit by deferring some items/links a footnote; we can improve/shorten things a bit further but this is an opportunity to advertise in particular with people less familiar with nim (eg too lazy to click on the website or who will be skimming through this on HN and have no idea about nim); ppl already familiar with nim will just skip this paragraph;

Overall, 18 lines out of 955 lines on "why use nim" is not an excessive amount IMO.

> also it will likely be better to write a dedicated article for Goodboy (The Gameboy Advance game developed by @exelotl)

a dedicated article on `Goodboy` would definitely be good to have (but it doesn't hurt to mention it here)

Reviewer 6: > Overall, 18 lines out of 955 lines on "why use nim" is not an excessive amount IMO.

18 lines isn't excessive. But 955 sure is :D

I've proof-read the article and got bored reading it before I reached 50% of the content.

* support for Apple silicon/M1, 32-bit RISC-V, armv8l, CROSSOS, improved support for NodeJS backend
* major improvements to the following modules: `system`, `math`, `random`, `json`, `jsonutils`, `os`, `typetraits`, `wrapnils`, `lists`, `hashes` including performance improvements
* deprecated a number of error prone or redundant mis-features

# Installing Nim 1.6
We recommend everyone to upgrade to 1.6:

## New users

Check out if your package manager already ships version 1.6 or
install it as described [here](https://nim-lang.org/install.html).

## Existing users

If you have installed a previous version of Nim using `choosenim`,
getting Nim 1.6 is as easy as:

```bash
choosenim update self
choosenim update stable
```

If you don't have `choosenim`, you can follow the same
[install link](https://nim-lang.org/install.html) as above.

## Building from source
```bash
git clone https://github.com/nim-lang/Nim
cd Nim
sh build_all.sh
```
The last command can be re-run after pulling new commits.

> Note that the `csources` repo used was changed to `csources_v1`, the new setup is designed to be forward and backward compatible.

Author: this and the other similar comments related to comma use are IMO controversial since we use this a lot so I'll keep things as is, unless more ppl care.

- `-d:nimLegacyMacrosCollapseSymChoice`
- `-d:nimLegacyParseQueryStrict`
- `-d:nimLegacyRandomInitRand`
- `-d:nimLegacyReprWithNewline`
- `-d:nimLegacySigpipeHandler`
- `-d:nimLegacyTypeMismatch`
- `-d:nimPreviewDotLikeOps`
- `-d:nimPreviewFloatRoundtrip`

# Major new features
With so many new features, pinpointing the most salient ones is a subjective exercise,
but here are a select few:

## `iterable[T]`
The `iterable[T]` type class was added to match called iterators,
which solves a number of long-standing issues related to iterators.
Example:
```nim
iterator iota(n: int): int =
  for i in 0..<n: yield i

# previously, you'd need `untyped`, which caused other problems such as lack
# of type inference, overloading issues, and MCS.
template sumOld(a: untyped): untyped = # no type inference possible
  var result: typeof(block:(for ai in a: ai))
  for ai in a: result += ai
  result

assert sumOld(iota(3)) == 0 + 1 + 2

# now, you can write:
template sum[T](a: iterable[T]): T =
  # `template sum(a: iterable): auto =` would also be possible
  var result: T
  for ai in a: result += ai
  result

assert sum(iota(3)) == 0 + 1 + 2 # or `iota(3).sum`

```

> In particular iterable arguments can now be used with MCS, example:

Author: ya, i really wish we'd embrace the more common term UFCS (even if it's approximation), but ...

> In particular iterable arguments can now be used with MCS, example:

Reviewer 7: "method call syntax" and "dot call syntax" are fine, no acronyms required.

- `addInt` now supports unsigned integers.

Compatibility notes:
- `system.delete` had surprising behavior when the index passed to it was out of
  bounds (it would delete the last entry then). Compile with `-d:nimStrictDelete` so
  that an index error is produced instead. Be aware however that your code might depend on
  this quirky behavior so a review process is required on your part before you can
  use `-d:nimStrictDelete`. To make this review easier, use the `-d:nimAuditDelete`
  switch, it pretends that `system.delete` is deprecated so that it is easier to see
  where it was used in your code.
  `-d:nimStrictDelete` will become the default in upcoming versions.
- `cuchar` is now deprecated as it aliased `char` where arguably it should have aliased `uint8`.
  Please use `char` or `uint8` instead.
- `repr` now doesn't insert trailing newlines; the previous behavior was very inconsistent,
  see [#16034](https://github.com/nim-lang/Nim/pull/16034). Use `-d:nimLegacyReprWithNewline` for the previous behavior. `repr` now also
  renders ASTs correctly for user defined literals, setters, `do`, etc.
- Deprecated `any`. See RFC [#281](https://github.com/nim-lang/RFCs/issues/281).

> - The unary slice `..b` was deprecated, use `0..b` instead.

Author: controversial

## `std/math`
- Added `almostEqual` for comparing two float values using a machine epsilon.
- Added `clamp` which allows using a `Slice` to clamp to a value.
- Added `ceilDiv` for round up integer division.
- Added `isNaN`.
- Added `copySign`.
- Added `euclDiv` and `euclMod`.
- Added `signbit`.

> - Added `frexp` overload procs. Deprecated `c_frexp`, use `frexp` instead.

Author: ditto

## `std/json`, `std/jsonutils`
- `std/jsonutils` now serializes/deserializes holey enums as regular enums (via `ord`) instead of as strings.
  Use `-d:nimLegacyJsonutilsHoleyEnum` for a transition period. `toJson` now serializes `JsonNode`
  as is via reference (without a deep copy) instead of treating `JsonNode` as a regular ref object,
  this can be customized via `jsonNodeMode`.
- `std/json` and `std/jsonutils` now serialize `NaN`, `Inf`, `-Inf` as strings, so that
  `%[NaN, -Inf]` is the string `["nan","-inf"]` instead of `[nan,-inf]` which was invalid JSON.
- `std/json` can now handle integer literals and floating point literals of
  arbitrary length and precision.
  Numbers that do not fit the underlying `BiggestInt` or `BiggestFloat` fields are
  kept as string literals and one can use external BigNum libraries to handle these.
  The `parseFloat` family of functions also has now optional `rawIntegers` and
  `rawFloats` parameters that can be used to enforce that all integer or float
  literals remain in the "raw" string form so that client code can easily treat
  small and large numbers uniformly.
- Added `BackwardsIndex` overload for `JsonNode`.

> - `json.%`,`json.to`, `jsonutils.formJson`,`jsonutils.toJson` now work with `uint|uint64`

Author: thanks, good catch

- Added optional `followSymlinks` argument to `setFilePermissions`.
- Added a simpler to use `io.readChars` overload.
- Added `socketstream` module that wraps sockets in the stream interface.
- Added experimental `linenoise.readLineStatus` to get line and status (e.g. ctrl-D or ctrl-C).

## Environment variable handling
- empty environment variable values are now supported across OS's and backends.
- environment variable APIs now work in multithreaded scenarios, by delegating to direct OS calls
  instead of trying to keep track of the environment.
- `putEnv`, `delEnv` now work at CT.
- NodeJS backend now supports osenv: `getEnv`, `putEnv`, `envPairs`, `delEnv`, `existsEnv`.

Compatibility notes:
- `std/os`: `putEnv` now raises if the 1st argument contains a `=`.

## POSIX
- On POSIX systems, the default signal handlers used for Nim programs (it's
  used for printing the stacktrace on fatal signals) will now re-raise the
  signal for the OS default handlers to handle.
  This lets the OS perform its default actions, which might include core
  dumping (on select signals) and notifying the parent process about the cause
  of termination.

> - On POSIX systems, we now ignore `SIGPIPE` signals, use `-d:nimLegacySigpipeHandler`

Author: ditto

- CI now supports batching (making Windows CI 2.3X faster).
- Sets now uses the optimized `countSetBits` proc, see PR [#17334](https://github.com/nim-lang/Nim/pull/17334).

## Debugging
- You can now enable/disable VM tracing in user code via `vmutils.vmTrace`.
- `koch tools` now builds `bin/nim_dbg` which allows easy access to a debug version of Nim without recompiling.
- Added new module `compiler/debugutils` to help with debugging Nim compiler.
- Renamed `-d:nimCompilerStackraceHints` to `-d:nimCompilerStacktraceHints` and used it in more contexts;
  this flag which works in tandem with `--stackTraceMsgs` to show user code context in compiler stacktraces.

## Type system
- `typeof(voidStmt)` now works and returns `void`.
- `enum` values can now be overloaded. This needs to be enabled
  via `{.experimental: "overloadableEnums".}`. We hope that this feature allows for the
  development of more fluent (less ugly) APIs. See RFC [#373](https://github.com/nim-lang/RFCs/issues/373).
  for more details.
- A type conversion from one `enum` type to another now produces an `[EnumConv]` warning.
  You should use `ord` (or `cast`, but the compiler won't help, if you misuse it) instead.

>   ```

Reviewer 9: ```suggestion
  ```nim
```

  type A = enum a1, a2
  type B = enum b1, b2
  echo a1.B # produces a warning
  echo a1.ord.B # produces no warning
  ```
- A dangerous implicit conversion to `cstring` now triggers a `[CStringConv]` warning.
  This warning will become an error in future versions! Use an explicit conversion
  like `cstring(x)` in order to silence the warning.
- There is a new warning for *any* type conversion to `enum` that can be enabled via
  `.warning[AnyEnumConv]:on` or `--warning:AnyEnumConv:on`.
- Reusing a type name in a different scope now works, refs [#17710](https://github.com/nim-lang/Nim/pull/17710).
- Fixed implicit and explicit generics in procedures, refs [#18808](https://github.com/nim-lang/Nim/pull/18808).

## New-style concepts
Example:
```nim
type
  Comparable = concept # no T, an atom
    proc cmp(a, b: Self): int
```

> See PR [#15251](https://github.com/nim-lang/Nim/pull/15251) for details.

Reviewer 10: This PR isn't very informative, maybe it would be better to link to the RFC? https://github.com/nim-lang/RFCs/issues/168

Reviewer 10: Could maybe include a sentence like: "The new design is easier to understand, faster to compile, and will allow future Nim versions to perform typechecking on generics at their point of declaration."

Author: there's also potential loss of flexibility (proc pragmas, matching calls that could be via MCS or otherwise) and the feature lacks testing; "easier to understand" has caveats too (eg not immediately obvious whether proc matches other routines etc)

I've nevertheless added a link to RFC and also this sentence:
```
The new design does not rely on `system.compiles` and may compile faster.
```

- Added `--spellSuggest` to show spelling suggestions on typos.
- Added `--processing:dots|filenames|off` which customizes `hintProcessing`;
  `--processing:filenames` shows which include/import modules are being compiled as an import stack.
- `FieldDefect` messages now shows discriminant value + lineinfo, in all backends (C, JS, VM)
- Added `--hintAsError` with similar semantics as `--warningAsError`.
- Added `--unitsep:on|off` to control whether to add ASCII unit separator `\31` before a newline
 for every generated message (potentially multiline), so tooling can tell when messages start and end.
- Added `--filenames:abs|canonical|legacyRelProj` which replaces `--listFullPaths:on|off`
- `--hint:all:on|off` is now supported to select or deselect all hints; it
  differs from `--hints:on|off` which acts as a (reversible) gate.
  Likewise with `--warning:all:on|off`.
- The style checking of the compiler now supports a `--styleCheck:usages` switch. This switch
  enforces that every symbol is written as it was declared, not enforcing
  the official Nim style guide. To be enabled, this has to be combined either
  with `--styleCheck:error` or `--styleCheck:hint`.

> - Type mismatch errors now show more context, use `-d:nimLegacyTypeMismatch` for previous

Author: ditto

- There is a new `cast` section `{.cast(uncheckedAssign).}: body` that disables some
  compiler checks regarding `case objects`. This allows serialization libraries
  to avoid ugly, non-portable solutions. See RFC [#407](https://github.com/nim-lang/RFCs/issues/407).
  for more details.

## Tools
- Major improvements to `nimgrep`, see PR [#15612
](https://github.com/nim-lang/Nim/pull/15612).
- `fusion` is now un-bundled from Nim, `./koch fusion` will
  install it via Nimble at a fixed hash.
- `testament`: added `nimoutFull: bool` spec to compare full output of compiler
  instead of a subset; many bugfixes to testament.

## Misc/cleanups
- Deprecated `TaintedString` and `--taintmode`.
- Deprecated `--nilseqs` which is now a noop.
- Added `-d:nimStrictMode` in CI in several places to ensure code doesn't have certain hints/warnings.
- Removed `.travis.yml`, `appveyor.yml.disabled`, `.github/workflows/ci.yml.disabled`.
- `[skip ci]` now works in azure and CI pipelines, see detail in PR [#17561](https://github.com/nim-lang/Nim/pull/17561).

# Footnotes
Tested on a 2.3 GHz 8-Core Intel Core i9, 2019 macOS 11.5 with 64GB RAM.

> * [1] command used: `nim c -o:/tmp/main -d:danger --eval:'echo "hello world"'`

Reviewer 11: ```suggestion
* [1] command used: `nim c -o:/tmp/main -d:danger --eval:'echo "hello world"'`.
Note that the Nim result is without optimizing for size (via `--opt:size`), without stripping (`--passL:-s`), and without using full link-time optimization (`--passC:-flto`).
It's also fairly straightforward to create a statically linked 5 kB "hello world" binary via musl, and you can get down to ~150 bytes with some extra tricks.
```

I think it's worth adding a couple of sentences here, as binary size is often a focus in reddit/HN comment threads. Feel free to adapt.

I've avoided `-d:strip` and `-d:lto` here, as they don't currently work reliably on every platform IIRC (macOS?).

Author: @ee7 see commit `update "small binaries" section` ; 

I didnt' mention `and you can get down to ~150 bytes with some extra tricks.` as I don't see where are those extra tricks mentioned; in particular https://irclogs.nim-lang.org/07-07-2020.html#12:31:34 mentions:
> and if I disregard echo and use c_printf from system/ansi_c it's 5112 bytes

do you have more detail?

@Yardanico can you please confirm the 5K numbers for static linking with musl? on osx (without musl) i get the quoted 49K

Reviewer 3: Sure, I'll check those numbers on Linux and Windows again to see how small will it be

Reviewer 3: @timotheecour it's better to discuss this matter in real-time chats I think, but here are the results for Zig 0.8.1 + Nim 1.5.1 (6bb32da4aecbe80a9cc0cce90322c0e741ee8b85):
Nim source file:
```nim
echo "Hello, World!"
```

Compilation process:
```
$ nim c --os:any -d:posix -d:noSignalHandler --cc:clang --clang.exe="zigcc.sh" --clang.linkerexe="zigcc.sh" -d:danger --gc:arc -d:useMalloc --panics:on --passL:"-target x86_64-linux-musl -flto" --opt:size hello.nim
Hint: used config file '/home/dian/nim/config/nim.cfg' [Conf]
Hint: used config file '/home/dian/nim/config/config.nims' [Conf]
...................................................
Hint:  [Link]
Hint: gc: arc; opt: size; options: -d:danger
23293 lines; 0.350s; 25.414MiB peakmem; proj: /home/dian/test_bin_size/hello.nim; out: /home/dian/test_bin_size/hello [SuccessX]

$ strip -s -R .comment -R .note -R .note.ABI-tag hello

$ wc hello
  14  166 7296 hello
```

`zigcc.sh` is a shell script with the contents:
```sh
#!/usr/bin/env sh
zig cc $@
```
It's needed because Nim expects a name of the compiler binary and doesn't accept commands with spaces like `zig cc`.

I've also done a couple of tests with a few musl gcc toolchains, but the binary size with them is around 13KB even after stripping.

P.S.: Not sure if we should really specify the whole process for building a small binary, it'll be quite cumbersome to follow.

Reviewer 3: Without the `--os:any -d:posix` trickery the file size is 16976 bytes.

Reviewer 11: > Result - a fully static binary with the size of 7296 bytes.

@Yardanico I get 4848 bytes for a `echo "Hello, World!` binary on a similar setup (x86_64; Arch Linux; Nim 1.5.1 recent commit; Zig 0.8.1).

It looks like you forgot this option: `--passC:'-flto -target x86_64-linux-musl'`.

The same thing, as a script:

```sh
#!/usr/bin/env sh
set -e
path_bin='./hello'

# Create source file
path_src="${path_bin}.nim"
rm -f "${path_src}"
echo 'echo "Hello, World!"' > "${path_src}"

# Compile
nim c -f \
  -d:danger --opt:size --gc:arc --panics:on \
  -d:useMalloc --os:any -d:posix -d:noSignalHandler \
  --cc=clang --clang.exe='zigcc.sh' --clang.linkerexe='zigcc.sh' \
  --passC:'-flto -target x86_64-linux-musl' \
  --passL:'-flto -target x86_64-linux-musl' \
  --skipUserCfg --skipParentCfg --skipProjCfg \
  "${path_bin}"

# Strip
strip --strip-all --remove-section=.comment "${path_bin}"

# Print some information about the binary
printf "\n"
file "${path_bin}"
byte_count="$(wc --bytes ${path_bin} | cut -d' ' -f1)"
echo "${byte_count} bytes"

# Show that the executable actually works
printf "\nRunning executable...\n"
"${path_bin}"
```

```
./hello: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, stripped
4848 bytes

Running executable...
Hello, World!
```

* [2] commands used:

>   for nim: `nim c --forceBuild compiler/nim`
>   for rust: `./x.py build`, see also https://www.reddit.com/r/rust/comments/76jq7h/long_time_to_compile_rustc/
>   for gcc: see https://unix.stackexchange.com/questions/421822/how-long-does-it-take-to-compile-gcc-7-3-0
>   https://solarianprogrammer.com/2016/10/07/building-gcc-ubuntu-linux/
>   for clang: see https://quuxplusone.github.io/blog/2018/04/16/building-llvm-from-source/
>   for go: `./make.bash`

Reviewer 8: ```suggestion
  for Nim: `nim c --forceBuild compiler/nim`
  for Rust: `./x.py build`, see also https://www.reddit.com/r/rust/comments/76jq7h/long_time_to_compile_rustc/
  for gcc: see https://unix.stackexchange.com/questions/421822/how-long-does-it-take-to-compile-gcc-7-3-0
  https://solarianprogrammer.com/2016/10/07/building-gcc-ubuntu-linux/
  for clang: see https://quuxplusone.github.io/blog/2018/04/16/building-llvm-from-source/
  for Go: `./make.bash`
```

Reviewer 3: Also `GCC` and `Clang`

Reviewer 3: ```suggestion
  for nim: `nim c --forceBuild compiler/nim`
  for rust: `./x.py build`, see also https://www.reddit.com/r/rust/comments/76jq7h/long_time_to_compile_rustc/
  for GCC: see https://unix.stackexchange.com/questions/421822/how-long-does-it-take-to-compile-gcc-7-3-0
  https://solarianprogrammer.com/2016/10/07/building-gcc-ubuntu-linux/
  for Clang: see https://quuxplusone.github.io/blog/2018/04/16/building-llvm-from-source/
  for go: `./make.bash`
```

in addition to the comment by konsumlamm

# now, you can write:
template sum[T](a: iterable[T]): T =
  # `template sum(a: iterable): auto =` would also be possible
  var result: T
  for ai in a: result += ai
  result

assert sum(iota(3)) == 0 + 1 + 2 # or `iota(3).sum`

```
In particular iterable arguments can now be used with the method call syntax. For example:
```nim
import std/[sequtils, os]
echo walkFiles("*").toSeq # now works
```
See PR [#17196](https://github.com/nim-lang/Nim/pull/17196) for additional details.

## Strict effects
The effect system was refined and there is a new `.effectsOf` annotation that does
explicitly what was previously done implicitly. See the [manual](https://nim-lang.github.io/Nim/manual.html#effect-system-effectsof-annotation) for more details.
To write code that is portable with older Nim versions, use this idiom:

```nim
when defined(nimHasEffectsOf):
  {.experimental: "strictEffects".}
else:
  {.pragma: effectsOf.}

proc mysort(s: seq; cmp: proc(a, b: T): int) {.effectsOf: cmp.}
```

To enable the new effect system, compile with `--experimental:strictEffects`. See also [#18777](https://github.com/nim-lang/Nim/pull/18777) and RFC [#408](https://github.com/nim-lang/RFCs/issues/408).

> 

Reviewer 12: I would link somehow to https://en.wikipedia.org/wiki/Side_effect_(computer_science)

Some people dont know what is a side effect in programming.

Author: strictEffects and effectsOf is more general than sideeffects though, and different languages have different definitions of what a side effect is (eg whether they're allowed through params); i think there is enough context here (including a ref to manual is provided)

Reviewer 7: Please @juancarlospaco stop the Wikipedia links, they are always terrible. Wikipedia is never a credible source and we don't need users who don't know about web search engines...

Reviewer 12: Feel free to edit Wikipedia if you dont like it.

I think Nim should reach a point were new users can learn programming with Nim,
if they have go to thru Python or JavaScript first to learn, then we lost them already...

Reviewer 7: We have reached that point already and you haven't noticed it. You can learn Nim programming without having to learn Python first and without reading Wikipedia articles.

echo weirdTarget # or `os.weirdTarget`
```

Added a new module `std/importutils`, and an API `privateAccess`, which allows access
to private fields for an object type in the current scope.

Example:
```nim
import times
from std/importutils import privateAccess
block:
  let t = now()
  # echo t.monthdayZero # Error: undeclared field: 'monthdayZero' for type times.DateTime
  privateAccess(typeof(t)) # enables private access in this scope
  echo t.monthdayZero # ok
```

See PR [#17706](https://github.com/nim-lang/Nim/pull/17706) for additional details.

## `nim --eval:cmd`
Added `nim --eval:cmd` to evaluate a command directly, e.g.: `nim --eval:"echo 1"`.
It defaults to `e` (nimscript) but can also work with other commands, e.g.:
```bash
find . | nim r --eval:'import strutils; for a in stdin.lines: echo a.toUpper'
```

```bash
# use as a calculator:
nim --eval:'echo 3.1 / (1.2+7)'
# explore a module's APIs, including private symbols:
nim --eval:'import os {.all.}; echo weirdTarget'
# use a custom backend:
nim r -b:js --eval:"import std/jsbigints; echo 2'big ** 64'big"
```
See PR [#15687](https://github.com/nim-lang/Nim/pull/15687) for more details.

> 

Reviewer 12: Mention that quotes are required for `--eval:`.

Author: i don't think it's needed as it's implied by the fact that it's an argument to a shell cmd; if more ppl agree with need for mentioning quotes are needed, i could reconsider

Example:
```nim
import std/jsbigints
assert 2'big ** 65'big == 36893488147419103232'big
echo 0xdeadbeef'big shl 4'big # 59774856944n
```

## New `std/sysrand` module
Cryptographically secure pseudorandom number generator,
allows generating random numbers from a secure source provided by the operating system.
Example:
```nim
import std/sysrand
assert urandom(1234) != urandom(1234) # unlikely to fail in practice
```
See PR [#16459](https://github.com/nim-lang/Nim/pull/16459).

## New module: `std/tempfiles`
Allows creating temporary files and directories, see PR [#17361](https://github.com/nim-lang/Nim/pull/17361) and followups.
```nim
import std/tempfiles
let tmpPath = genTempPath("prefix", "suffix.log", "/tmp/")
# tmpPath looks like: /tmp/prefixpmW1P2KLsuffix.log

let dir = createTempDir("tmpprefix_", "_end")
# created dir looks like: getTempDir() / "tmpprefix_YEl9VuVj_end"

let (cfile, path) = createTempFile("tmpprefix_", "_end.tmp")
# path looks like: getTempDir() / "tmpprefix_FDCIRZA0_end.tmp"
cfile.write "foo"
cfile.setFilePos 0
assert readAll(cfile) == "foo"
close cfile
assert readFile(path) == "foo"
```

> 

Reviewer 12: This module still needs to do automatic clean-ups, maybe mention that?.

Author: not sure what you mean; the example illustrates one needs to write `close cfile`
future work could add a `tmpFileScoped`, but we don't have it right now

Reviewer 12: Original PR says that it still lacks destructors to do auto-clean up,
Python documentation also mentions which ones do auto-clean up, and which ones does not,
but you are right, maybe that should be on the documentation itself.

An important use case is to enable dynamic fields without affecting the built-in `.` operator, e.g. for
`std/jsffi`, `std/json`, `pkg/nimpy`. Example:
```nim
import std/json
template `.?`(a: JsonNode, b: untyped{ident}): JsonNode =
  a[astToStr(b)]
let j = %*{"a1": {"a2": 10}}
assert j.?a1.?a2.getInt == 10
```

## Block arguments now support optional parameters
This solves a major pain point for routines accepting block parameters, see PR [#18631](https://github.com/nim-lang/Nim/pull/18631) for details:
```nim
template fn(a = 1, b = 2, body) = discard
fn(1, 2): # already works
  bar
fn(a = 1): # now works
  bar
```
Likewise with multiple block arguments via `do`:
```nim
template fn(a = 1, b = 2, body1, body2) = discard
fn(a = 1): # now works
  bar1
do:
  bar2
```

# Other new features

## New and deprecated modules
The following modules were added (they are discussed in the rest of the text):
- `std/enumutils`
- `std/genasts`
- `std/importutils`
- `std/jsbigints`
- `std/jsfetch`
- `std/jsformdata`
- `std/jsheaders`
- `std/packedsets`
- `std/setutils`
- `std/socketstreams`
- `std/strbasics`
- `std/sysrand`
- `std/tasks`
- `std/tempfiles`

> - `std/vmutils`

Reviewer 12: I would add a `std/jsfetch` example, that also uses `then()` and `catch()`,
maybe just copy the one from the documentation.

Also an example for `std/tasks` would be nice, can also be copied from documentation.

Author: > I would add a std/jsfetch example, that also uses then() and catch(),

good idea, done

Author: > Also an example for std/tasks would be nice, can also be copied from documentation.

i adapted / shortened it; @xflywind please review

Author: > Example for std/genasts too, because is mentioned as "alternative" for quote do.

done

Reviewer 9: LGTM

* no need for makefiles, cmake, configure or other build scripts, thanks to CTFE and dependency tracking [3]
* target any platform with a C compiler: [Android and iOS](https://github.com/pragmagic/godot-nim#made-with-godot-nim),
  [embedded systems](https://github.com/elcritch/nesper), [micro-controllers](https://forum.nim-lang.org/t/7731),
  [WASM](https://forum.nim-lang.org/t/4779), Nintendo Switch, [Game Boy Advance](https://forum.nim-lang.org/t/8375).
* zero-overhead interop lets you reuse code in C, C++ (including templates, [C++ STL](https://clonkk.github.io/nim-cppstl/cppstl.html)), JS, Objective-C, Python (via [nimpy](https://github.com/yglukhov/nimpy))
* built-in [documentation generator](https://nim-lang.github.io/Nim/system.html) that understands Nim code and runnable examples that stay in sync

Last but not least, macros let you manipulate/generate code at compile time instead of relying on code generators,
enabling writing DSLs and language extensions in user code. Typical examples include implementing

> Python f-strings, optional chaining, command line generators, react-like Single Page Apps, protobuf serialization and binding generators [4].

Reviewer 8: ```suggestion
Python-like f-strings, optional chaining, command line generators, react-like Single Page Apps, protobuf serialization and binding generators [4].
```

* significant improvements in error messages, showing useful context
* documentation generation logic and documentation, in particular `runnableExamples` now works in more contexts
* JS, VM and nimscript backend are more consistent with C backend, allowing more modules to work with those backends, including the imports from `std/prelude`; the test suite now standardizes on testing stdlib modules on each major backend (C, JS, VM).

> * support for Apple silicon/M1, 32-bit RISC-V, armv8l, CROSSOS, improved support for NodeJS backend

Reviewer 3: why duplicate this in two places?

Author: this is a short summary of the most salient points; the rest of the text has more details including links to PRs in some cases as well as grouping in semantic sections

- Empty environment variable values are now supported across OS's and backends.
- Environment variable APIs now work in multithreaded scenarios, by delegating to direct OS calls
  instead of trying to keep track of the environment.
- NodeJS backend now supports `osenv`: `getEnv`, `putEnv`, `envPairs`, `delEnv`, `existsEnv`.

> - `osenv` now works in all backends (c,cpp,vm,nims,js with -d:nodejs), see PR [#18615](https://github.com/nim-lang/Nim/pull/18615).

Reviewer 9: This PR may not be backported to Nim 1.6

## OS-specific notes
- Support for Apple silicon/M1.

> - Support for 32-bit RISC-V, refs [#16231](https://github.com/nim-lang/Nim/pull/16231).

Reviewer 3: how are M1, 32-bit RISC-V and armv8l OS specific?

Author: changed title to: `## OS/platform-specific notes`

@Yardanico feel free to suggest alternatives
